{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Create MHC + CH status dataframe  \n",
    "- Barbara Walkowiak (bw450)\n",
    "- 2024-06-19, modified 2024-09-07\n",
    "\n",
    "- This script is used to generate a dataframe with MHC and CH data for all individuals screened for CH to be used for later analysis (main + supplementary figures)\n",
    "\n",
    "- Dataframe contents\n",
    "    - Person ID for everyone screened for CH\n",
    "    - their MHC genotype (df is filtered to only include people w/ confidently imputed MHC genotype)\n",
    "    - MHC-associated parameters (heterozygosity status for each locus, overall, nr of alleles) \n",
    "    - CH status (gene_var = variant they carry, or NaN if they don't carry a variant) \n",
    "    - VAF for the variant they carry (or NaN if no CH) + read depth (total) + var read depth (nr reads with the variant)\n",
    "    - MHCI-based score for each variant (40 examined variants for now)\n",
    "    - MHCII-based score for each variant (40 examined variants for now) \n",
    "    - median score for each variant (the same for everyone because there is only one median score for a given variant)\n",
    "    - group for each of the 40 variants (top binding vs bottom binding; population split in half) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Set up"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "# IMPORTS\n",
    "\n",
    "import random\n",
    "import math\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.ticker as plticker\n",
    "from matplotlib.ticker import LinearLocator, FormatStrFormatter\n",
    "from matplotlib.patches import Polygon\n",
    "from matplotlib.patches import Patch\n",
    "from matplotlib.lines import Line2D\n",
    "from matplotlib import cm\n",
    "import scipy.special\n",
    "from scipy import integrate\n",
    "import scipy.integrate as it\n",
    "from scipy.interpolate import interp1d\n",
    "from scipy.stats import kde\n",
    "import copy\n",
    "import glob, os\n",
    "import re\n",
    "# from sklearn import datasets, linear_model\n",
    "import pandas as pd\n",
    "from decimal import *\n",
    "from operator import itemgetter    \n",
    "from collections import OrderedDict\n",
    "import timeit\n",
    "import time \n",
    "import csv\n",
    "import seaborn as sns \n",
    "import scipy as sp\n",
    "from sklearn.preprocessing import LabelEncoder\n",
    "from sklearn.decomposition import PCA\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from itertools import combinations\n",
    "from scipy.stats import kruskal\n",
    "from scipy.stats import mannwhitneyu\n",
    "from matplotlib.backends.backend_pdf import PdfPages\n",
    "import statsmodels.api as sm\n",
    "import plotly.express as px\n",
    "import kaleido\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "# specify font for plotting \n",
    "plt.rcParams.update({'font.sans-serif':'Helvetica'})\n",
    "\n",
    "# stop printing warnings \n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "# get current date \n",
    "timestr = time.strftime(\"%Y%m%d\") "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# GOAL: df with everyone screened for CH: ID, MHC I / II, CH status, age"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Determine IDs of participants who were screened for CH\n",
    "- here, find IDs of EVERYONE screened to determine if they had CH or not"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of data batches screened for CH: 51\n",
      "Number of people who were screened for CH: 454800\n",
      "Person_IDs: [1876197, 1866600, 1891748, 1874461]\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# import indexes screened in each batch\n",
    "folder_path = '/Users/barbarawalkowiak/Desktop/msc_thesis/data/ukb_hotspot_calls/batch_ids'  # Path to ukb files from Hamish \n",
    "\n",
    "# Identify all tsv files in the folder\n",
    "files_ids = glob.glob(os.path.join(folder_path, '*.tsv'))  \n",
    "\n",
    "# read each file\n",
    "indexes = {}\n",
    "\n",
    "# read each CSV \n",
    "for file in files_ids:\n",
    "    \n",
    "    id = pd.read_csv(file, sep = '\\t')\n",
    "    id = id.rename(columns={'batch ID': 'sample_ID'})\n",
    "    id['sample_ID'] = id['sample_ID'].str.split('_', n = 1).str[0]\n",
    "    indexes[file] = id\n",
    "\n",
    "for file_name, df in indexes.items():\n",
    "    \n",
    "    variable_name = file_name.split('/')[8].split('.')[0] # get rid of tsv \n",
    "    globals()[variable_name] = df \n",
    "\n",
    "print('Number of data batches screened for CH:', len(indexes)) # should be 51, as 51 batches were screened\n",
    "\n",
    "indexes = { k: v.set_index('sample_ID') for k, v in indexes.items()}\n",
    "\n",
    "indexes_df = pd.concat(indexes, axis=1)\n",
    "indexes_df.columns = indexes_df.columns.droplevel(-1)\n",
    "indexes_df = indexes_df.reset_index()\n",
    "indexes_df.head()\n",
    "\n",
    "indexes_list = indexes_df['sample_ID'].tolist()\n",
    "indexes_list = [int(id) for id in indexes_list]\n",
    "print('Number of people who were screened for CH:', len(indexes_list))\n",
    "print('Person_IDs:', indexes_list[0:4]) # check that Person IDs are in the correct format (numeric)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load the CH variant data\n",
    "- add CH status (CH-positive if >= 2 reads with one of the pre-defined hotspot positions)\n",
    "- presence of > 1 hotspot mutation is allowed (i.e., possible for someone to carry 2 or more variants)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Minimum number of reads in the sample: 2\n",
      "Minimum number of reads in the sample (TP53): 3\n",
      "Number of samples with >=2 reads with a change in CH-hotspot POSITION: 11310\n",
      "Number of individuals with >=2 reads with a change in CH-hotspot POSITION: 10931\n",
      "Number of samples with >=2 reads with a change in TP53-hotspot POSITION: 226\n",
      "Number of individuals with >=2 reads with a change in TP53-hotspot POSITION: 226\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# Import all the data (all calls: include singletons and higher read numbers, for all changes at a position affected in CH)\n",
    "\n",
    "# read in the data (files received from Hamish MacGregor)\n",
    "df_all_variants = pd.read_csv('/Users/barbarawalkowiak/Desktop/msc_thesis/data/ukb_healthy_curated_variants_not_excluding_multi_individuals.tsv', sep = '\\t') \n",
    "df_tp53_variants = pd.read_csv('/Users/barbarawalkowiak/Desktop/msc_thesis/data/TP53_putative_somatic.tsv', sep = '\\t') # separate file for TP53 calling \n",
    "\n",
    "# remove individuals with mCAs (most are in JAK2)\n",
    "df_all_variants = df_all_variants[df_all_variants['overlap'] == False]\n",
    "\n",
    "# this dataset has been filtered to only include individuals who have >= 2 reads corresponding to a given CH variant \n",
    "print('Minimum number of reads in the sample:', min(df_all_variants['var_depth']))\n",
    "print('Minimum number of reads in the sample (TP53):', min(df_tp53_variants['var_depth'])) # for TP53, using 3 reads because the prior on them being real CH and not an error is lower\n",
    "\n",
    "# print the number of SAMPLES and INDIVIDUALS with mutations in CH hotspot POSITIONS\n",
    "print('Number of samples with >=2 reads with a change in CH-hotspot POSITION:', df_all_variants.shape[0])\n",
    "print('Number of individuals with >=2 reads with a change in CH-hotspot POSITION:', len(df_all_variants.sample_ID.unique()))\n",
    "print('Number of samples with >=2 reads with a change in TP53-hotspot POSITION:', df_tp53_variants.shape[0])\n",
    "print('Number of individuals with >=2 reads with a change in TP53-hotspot POSITION:', len(df_tp53_variants.sample_ID.unique()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filtering: only include variants carried by more than 15 individuals (common enough)\n",
    "\n",
    "# TP53 variants: select only the two most common ones (require > 15 people who carry the variant)\n",
    "tp53_counts = pd.DataFrame(df_tp53_variants.varID.value_counts()).reset_index()\n",
    "tp53_hotspots = tp53_counts[tp53_counts['count'] > 15].varID.tolist()\n",
    "\n",
    "# subset to only include individuals positive for the common hotspots  \n",
    "df_tp53_hotspots = df_tp53_variants[df_tp53_variants['varID'].isin(tp53_hotspots)]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of CH variants identified: 76\n",
      "All CH variants identified in the dataset: ['DNMT3A L859F' 'DNMT3A L859L' 'DNMT3A P904L' 'DNMT3A P904Q'\n",
      " 'DNMT3A P904R' 'DNMT3A R320*' 'DNMT3A R320G' 'DNMT3A R320R'\n",
      " 'DNMT3A R326C' 'DNMT3A R326G' 'DNMT3A R326S' 'DNMT3A R598*'\n",
      " 'DNMT3A R598G' 'DNMT3A R598R' 'DNMT3A R729G' 'DNMT3A R729R'\n",
      " 'DNMT3A R729W' 'DNMT3A R736C' 'DNMT3A R736G' 'DNMT3A R736H'\n",
      " 'DNMT3A R736L' 'DNMT3A R736P' 'DNMT3A R736S' 'DNMT3A R771*'\n",
      " 'DNMT3A R771G' 'DNMT3A R771R' 'DNMT3A R882C' 'DNMT3A R882G'\n",
      " 'DNMT3A R882H' 'DNMT3A R882L' 'DNMT3A R882P' 'DNMT3A R882S'\n",
      " 'DNMT3A Y735C' 'DNMT3A Y735F' 'DNMT3A Y735S' 'GNB1 K57*' 'GNB1 K57E'\n",
      " 'IDH1 R132H' 'IDH1 R132L' 'IDH2 R140L' 'IDH2 R140P' 'IDH2 R140Q'\n",
      " 'IDH2 R172K' 'IDH2 R172M' 'JAK2 V617F' 'JAK2 V617I' 'KIT D816G'\n",
      " 'KIT D816N' 'KIT D816V' 'KRAS G12A' 'KRAS G12C' 'KRAS G12D' 'KRAS G12R'\n",
      " 'KRAS G12S' 'KRAS G12V' 'MPL W515*' 'MPL W515L' 'MPL W515S' 'NPM1 L287L'\n",
      " 'NPM1 W288G' 'NPM1 W288R' 'NRAS G12A' 'NRAS G12C' 'NRAS G12D' 'NRAS G12R'\n",
      " 'NRAS G12S' 'NRAS G12V' 'SF3B1 K666K' 'SF3B1 K666N' 'SF3B1 K700*'\n",
      " 'SF3B1 K700E' 'SRSF2 P95H' 'SRSF2 P95L' 'SRSF2 P95R'] ['TP53 R273H', 'TP53 R175H']\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>varID</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DNMT3A R882H</td>\n",
       "      <td>1860</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DNMT3A R882C</td>\n",
       "      <td>1109</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DNMT3A Y735C</td>\n",
       "      <td>873</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DNMT3A R736H</td>\n",
       "      <td>647</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DNMT3A P904L</td>\n",
       "      <td>516</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>NRAS G12R</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>KIT D816G</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>DNMT3A L859L</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>SF3B1 K666K</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>IDH2 R140P</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>74 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           varID  count\n",
       "0   DNMT3A R882H   1860\n",
       "1   DNMT3A R882C   1109\n",
       "2   DNMT3A Y735C    873\n",
       "3   DNMT3A R736H    647\n",
       "4   DNMT3A P904L    516\n",
       "..           ...    ...\n",
       "69     NRAS G12R      1\n",
       "70     KIT D816G      1\n",
       "71  DNMT3A L859L      1\n",
       "72   SF3B1 K666K      1\n",
       "73    IDH2 R140P      1\n",
       "\n",
       "[74 rows x 2 columns]"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# This dataframe includes variant annotation in the varID column \n",
    "\n",
    "# Check which variants were identified \n",
    "print('Number of CH variants identified:', len(df_all_variants['varID'].unique()) + len(tp53_hotspots))\n",
    "print('All CH variants identified in the dataset:', df_all_variants['varID'].sort_values().unique(), tp53_hotspots)\n",
    "\n",
    "# show a table which shows the number of cases identified for each variant (not including TP53 here - this has already been filtered)\n",
    "ch_var_counts = pd.DataFrame(df_all_variants.varID.value_counts()).reset_index()\n",
    "ch_var_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Exclude synonymous variants (which were called due to nucleotide-level difference)\n",
    "syn_variants = ['DNMT3A L859L', 'DNMT3A R320R', 'DNMT3A R598R', 'DNMT3A R729R', 'DNMT3A R771R', 'NPM1 L287L', 'SF3B1 K666K']\n",
    "df_syn_variants = df_all_variants[df_all_variants['varID'].isin(syn_variants)]\n",
    "# these are clearly not hotspots / drivers so can be excluded"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of common (>= 15 cases) CH variants identified in the dataset: 38\n",
      "All common (>= 15 cases) CH variants identified in the dataset: ['DNMT3A P904L' 'DNMT3A P904Q' 'DNMT3A P904R' 'DNMT3A R320*'\n",
      " 'DNMT3A R326C' 'DNMT3A R326G' 'DNMT3A R326S' 'DNMT3A R598*'\n",
      " 'DNMT3A R729G' 'DNMT3A R729W' 'DNMT3A R736C' 'DNMT3A R736G'\n",
      " 'DNMT3A R736H' 'DNMT3A R736L' 'DNMT3A R736S' 'DNMT3A R771*'\n",
      " 'DNMT3A R882C' 'DNMT3A R882H' 'DNMT3A R882L' 'DNMT3A R882P'\n",
      " 'DNMT3A R882S' 'DNMT3A Y735C' 'DNMT3A Y735F' 'DNMT3A Y735S' 'GNB1 K57E'\n",
      " 'IDH1 R132H' 'IDH2 R140Q' 'IDH2 R172K' 'JAK2 V617F' 'KRAS G12D'\n",
      " 'KRAS G12S' 'MPL W515L' 'NRAS G12D' 'SF3B1 K666N' 'SF3B1 K700E'\n",
      " 'SRSF2 P95H' 'SRSF2 P95L' 'SRSF2 P95R']\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_ID</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>position</th>\n",
       "      <th>end_position</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>depth</th>\n",
       "      <th>var_depth</th>\n",
       "      <th>VAF</th>\n",
       "      <th>variant_type</th>\n",
       "      <th>...</th>\n",
       "      <th>VARIANT_CLASS</th>\n",
       "      <th>varID</th>\n",
       "      <th>overlap</th>\n",
       "      <th>Sex_v0</th>\n",
       "      <th>Age.when.attended.assessment.centre_v0</th>\n",
       "      <th>first.diagnosed.cancer.date1</th>\n",
       "      <th>first.diagnosed.cancer.age1</th>\n",
       "      <th>first.diagnosed.cancer.type1</th>\n",
       "      <th>Date.of.death_v0</th>\n",
       "      <th>Age.at.death_v0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1000130</td>\n",
       "      <td>chr2</td>\n",
       "      <td>25234373</td>\n",
       "      <td>25234373</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>81</td>\n",
       "      <td>2</td>\n",
       "      <td>0.024691</td>\n",
       "      <td>SNV</td>\n",
       "      <td>...</td>\n",
       "      <td>SNV</td>\n",
       "      <td>DNMT3A R882H</td>\n",
       "      <td>False</td>\n",
       "      <td>Female</td>\n",
       "      <td>41</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1000559</td>\n",
       "      <td>chr2</td>\n",
       "      <td>25234374</td>\n",
       "      <td>25234374</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>56</td>\n",
       "      <td>2</td>\n",
       "      <td>0.035714</td>\n",
       "      <td>SNV</td>\n",
       "      <td>...</td>\n",
       "      <td>SNV</td>\n",
       "      <td>DNMT3A R882C</td>\n",
       "      <td>False</td>\n",
       "      <td>Male</td>\n",
       "      <td>68</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1000724</td>\n",
       "      <td>chr2</td>\n",
       "      <td>25240313</td>\n",
       "      <td>25240313</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>98</td>\n",
       "      <td>2</td>\n",
       "      <td>0.020408</td>\n",
       "      <td>SNV</td>\n",
       "      <td>...</td>\n",
       "      <td>SNV</td>\n",
       "      <td>DNMT3A R771*</td>\n",
       "      <td>False</td>\n",
       "      <td>Female</td>\n",
       "      <td>62</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1001048</td>\n",
       "      <td>chr2</td>\n",
       "      <td>25247629</td>\n",
       "      <td>25247629</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>88</td>\n",
       "      <td>2</td>\n",
       "      <td>0.022727</td>\n",
       "      <td>SNV</td>\n",
       "      <td>...</td>\n",
       "      <td>SNV</td>\n",
       "      <td>DNMT3A R326G</td>\n",
       "      <td>False</td>\n",
       "      <td>Female</td>\n",
       "      <td>43</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1001150</td>\n",
       "      <td>chr2</td>\n",
       "      <td>25247647</td>\n",
       "      <td>25247647</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>74</td>\n",
       "      <td>2</td>\n",
       "      <td>0.027027</td>\n",
       "      <td>SNV</td>\n",
       "      <td>...</td>\n",
       "      <td>SNV</td>\n",
       "      <td>DNMT3A R320*</td>\n",
       "      <td>False</td>\n",
       "      <td>Male</td>\n",
       "      <td>67</td>\n",
       "      <td>2012-12-04</td>\n",
       "      <td>72.5</td>\n",
       "      <td>C92</td>\n",
       "      <td>2013-06-12</td>\n",
       "      <td>73.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11492</th>\n",
       "      <td>6022096</td>\n",
       "      <td>chr2</td>\n",
       "      <td>25244214</td>\n",
       "      <td>25244214</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>113</td>\n",
       "      <td>2</td>\n",
       "      <td>0.017699</td>\n",
       "      <td>SNV</td>\n",
       "      <td>...</td>\n",
       "      <td>SNV</td>\n",
       "      <td>DNMT3A R598*</td>\n",
       "      <td>False</td>\n",
       "      <td>Male</td>\n",
       "      <td>41</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11493</th>\n",
       "      <td>6024372</td>\n",
       "      <td>chr9</td>\n",
       "      <td>5073770</td>\n",
       "      <td>5073770</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>20</td>\n",
       "      <td>2</td>\n",
       "      <td>0.100000</td>\n",
       "      <td>SNV</td>\n",
       "      <td>...</td>\n",
       "      <td>SNV</td>\n",
       "      <td>JAK2 V617F</td>\n",
       "      <td>False</td>\n",
       "      <td>Male</td>\n",
       "      <td>47</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11495</th>\n",
       "      <td>6025111</td>\n",
       "      <td>chr2</td>\n",
       "      <td>25244214</td>\n",
       "      <td>25244214</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>89</td>\n",
       "      <td>2</td>\n",
       "      <td>0.022472</td>\n",
       "      <td>SNV</td>\n",
       "      <td>...</td>\n",
       "      <td>SNV</td>\n",
       "      <td>DNMT3A R598*</td>\n",
       "      <td>False</td>\n",
       "      <td>Female</td>\n",
       "      <td>62</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11496</th>\n",
       "      <td>6025356</td>\n",
       "      <td>chr2</td>\n",
       "      <td>25247629</td>\n",
       "      <td>25247629</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>152</td>\n",
       "      <td>2</td>\n",
       "      <td>0.013158</td>\n",
       "      <td>SNV</td>\n",
       "      <td>...</td>\n",
       "      <td>SNV</td>\n",
       "      <td>DNMT3A R326S</td>\n",
       "      <td>False</td>\n",
       "      <td>Male</td>\n",
       "      <td>55</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11497</th>\n",
       "      <td>6025637</td>\n",
       "      <td>chr2</td>\n",
       "      <td>25240313</td>\n",
       "      <td>25240313</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>97</td>\n",
       "      <td>3</td>\n",
       "      <td>0.030928</td>\n",
       "      <td>SNV</td>\n",
       "      <td>...</td>\n",
       "      <td>SNV</td>\n",
       "      <td>DNMT3A R771*</td>\n",
       "      <td>False</td>\n",
       "      <td>Female</td>\n",
       "      <td>59</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10547 rows × 51 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       sample_ID chromosome  position  end_position ref alt  depth  var_depth  \\\n",
       "0        1000130       chr2  25234373      25234373   C   T     81          2   \n",
       "1        1000559       chr2  25234374      25234374   G   A     56          2   \n",
       "2        1000724       chr2  25240313      25240313   G   A     98          2   \n",
       "3        1001048       chr2  25247629      25247629   G   C     88          2   \n",
       "4        1001150       chr2  25247647      25247647   G   A     74          2   \n",
       "...          ...        ...       ...           ...  ..  ..    ...        ...   \n",
       "11492    6022096       chr2  25244214      25244214   G   A    113          2   \n",
       "11493    6024372       chr9   5073770       5073770   G   T     20          2   \n",
       "11495    6025111       chr2  25244214      25244214   G   A     89          2   \n",
       "11496    6025356       chr2  25247629      25247629   G   T    152          2   \n",
       "11497    6025637       chr2  25240313      25240313   G   A     97          3   \n",
       "\n",
       "            VAF variant_type  ... VARIANT_CLASS         varID  overlap  \\\n",
       "0      0.024691          SNV  ...           SNV  DNMT3A R882H    False   \n",
       "1      0.035714          SNV  ...           SNV  DNMT3A R882C    False   \n",
       "2      0.020408          SNV  ...           SNV  DNMT3A R771*    False   \n",
       "3      0.022727          SNV  ...           SNV  DNMT3A R326G    False   \n",
       "4      0.027027          SNV  ...           SNV  DNMT3A R320*    False   \n",
       "...         ...          ...  ...           ...           ...      ...   \n",
       "11492  0.017699          SNV  ...           SNV  DNMT3A R598*    False   \n",
       "11493  0.100000          SNV  ...           SNV    JAK2 V617F    False   \n",
       "11495  0.022472          SNV  ...           SNV  DNMT3A R598*    False   \n",
       "11496  0.013158          SNV  ...           SNV  DNMT3A R326S    False   \n",
       "11497  0.030928          SNV  ...           SNV  DNMT3A R771*    False   \n",
       "\n",
       "       Sex_v0 Age.when.attended.assessment.centre_v0  \\\n",
       "0      Female                                     41   \n",
       "1        Male                                     68   \n",
       "2      Female                                     62   \n",
       "3      Female                                     43   \n",
       "4        Male                                     67   \n",
       "...       ...                                    ...   \n",
       "11492    Male                                     41   \n",
       "11493    Male                                     47   \n",
       "11495  Female                                     62   \n",
       "11496    Male                                     55   \n",
       "11497  Female                                     59   \n",
       "\n",
       "       first.diagnosed.cancer.date1  first.diagnosed.cancer.age1  \\\n",
       "0                               NaN                          NaN   \n",
       "1                               NaN                          NaN   \n",
       "2                               NaN                          NaN   \n",
       "3                               NaN                          NaN   \n",
       "4                        2012-12-04                         72.5   \n",
       "...                             ...                          ...   \n",
       "11492                           NaN                          NaN   \n",
       "11493                           NaN                          NaN   \n",
       "11495                           NaN                          NaN   \n",
       "11496                           NaN                          NaN   \n",
       "11497                           NaN                          NaN   \n",
       "\n",
       "      first.diagnosed.cancer.type1 Date.of.death_v0 Age.at.death_v0  \n",
       "0                              NaN              NaN             NaN  \n",
       "1                              NaN              NaN             NaN  \n",
       "2                              NaN              NaN             NaN  \n",
       "3                              NaN              NaN             NaN  \n",
       "4                              C92       2013-06-12            73.0  \n",
       "...                            ...              ...             ...  \n",
       "11492                          NaN              NaN             NaN  \n",
       "11493                          NaN              NaN             NaN  \n",
       "11495                          NaN              NaN             NaN  \n",
       "11496                          NaN              NaN             NaN  \n",
       "11497                          NaN              NaN             NaN  \n",
       "\n",
       "[10547 rows x 51 columns]"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# look for variants with count >= 15 \n",
    "common_ch_variants = ch_var_counts[ch_var_counts['count'] >= 15].varID # identify variants present in more than 15 individuals \n",
    "common_ch_variants = [var for var in common_ch_variants if var not in syn_variants] # remove variants which are classified as synonymous \n",
    "df_common_variants = df_all_variants[df_all_variants['varID'].isin(common_ch_variants)]\n",
    "\n",
    "print('Number of common (>= 15 cases) CH variants identified in the dataset:', len(df_common_variants['varID'].unique()))\n",
    "print('All common (>= 15 cases) CH variants identified in the dataset:', df_common_variants['varID'].sort_values().unique())\n",
    "\n",
    "df_common_variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of samples with common CH variants (>=2 reads, non-TP53): 10547\n",
      "Number of individuals with common CH variants (>=2 reads, non-TP53): 10198\n",
      "Number of samples with common CH variants (>=3 reads, non-TP53): 5153\n",
      "Number of individuals with common CH variants (>=3 reads, non-TP53): 4978\n",
      "Total number of common variants to look at: 40\n",
      "Total number of samples with any of these variants: 10602\n"
     ]
    }
   ],
   "source": [
    "\n",
    "print('Number of samples with common CH variants (>=2 reads, non-TP53):', df_common_variants.shape[0])\n",
    "print('Number of individuals with common CH variants (>=2 reads, non-TP53):', len(df_common_variants.sample_ID.unique()))\n",
    "\n",
    "print('Number of samples with common CH variants (>=3 reads, non-TP53):', df_common_variants[df_common_variants['var_depth']>=3].shape[0])\n",
    "print('Number of individuals with common CH variants (>=3 reads, non-TP53):', len(df_common_variants[df_common_variants['var_depth']>=3].sample_ID.unique()))\n",
    "\n",
    "print('Total number of common variants to look at: 40')\n",
    "print('Total number of samples with any of these variants:', df_common_variants.shape[0] + df_tp53_hotspots.shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of CH-positive individuals: 10250\n",
      "Number of CH variants in the UKB: 10602\n"
     ]
    }
   ],
   "source": [
    "# merge dataframes so that you have everyone who carries a variant you intend to look at\n",
    "df_common_variants_all = pd.concat([df_common_variants, df_tp53_hotspots])\n",
    "print('Number of CH-positive individuals:', len(df_common_variants_all.sample_ID.unique().tolist()))\n",
    "print('Number of CH variants in the UKB:', df_common_variants_all.shape[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load MHC genotyping data \n",
    "- add MHC I and II genotyping data \n",
    "- add age data (this does matter as CH is age-dependent so may be helpful, and useful for regression)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of samples with imputed HLA: 488377\n",
      "Number of samples with confidently imputed MHC I: 443620\n",
      "Number of samples with confidently imputed MHC II: 378317\n",
      "Number of UKBB participants with correctly genotyped MHC class I and II: 346968\n",
      "Number of UKBB participants with correctly genotyped MHC class I and II: 346968\n"
     ]
    }
   ],
   "source": [
    "# add MHC genotype data to CH cases\n",
    "\n",
    "# path to file\n",
    "file_hla = \"/Users/barbarawalkowiak/Desktop/msc_thesis/data/UKBB/ukb_typed_hla_with_ids_07_threshold_processed.txt\"\n",
    "\n",
    "# get the header \n",
    "header = pd.read_csv(file_hla, sep='\\t', nrows=1, header=None).values.tolist()\n",
    "head = [item for sublist in header for item in sublist]\n",
    "\n",
    "# load the dataframe \n",
    "df = pd.read_csv(file_hla, skiprows = 1, sep = ' ', header = None)\n",
    "\n",
    "# add columns \n",
    "df.columns = head \n",
    "\n",
    "# subset data for HLA-I and HLA-II class alleles \n",
    "df_hla1 = df.filter(regex='^(Person_|A_|B_|C_)') # 488377 cases \n",
    "df_hla2 = df.filter(regex='^(Person_|D)') # 488377 cases \n",
    "\n",
    "# clean up the dataset (filtering according to Krishna et al., 2024 Science: remove people in whom not all alleles were genotyped with sufficient confidence)\n",
    "df_clean_hla1 = df_hla1[~df_hla1.isin([0.5]).any(axis=1)] # exclude people for whom the genotype is missing \n",
    "df_clean_hla1 = df_clean_hla1[~df_clean_hla1.isin([1.5]).any(axis=1)] # exclude people for whom the genotype is missing\n",
    "df_clean_hla2 = df_hla2[~df_hla2.isin([0.5]).any(axis=1)] # exclude people for whom the genotype is missing\n",
    "df_clean_hla2 = df_clean_hla2[~df_clean_hla2.isin([1.5]).any(axis=1)] # exclude people for whom the genotype is missing\n",
    "\n",
    "print(\"Number of samples with imputed HLA:\", df_hla1.shape[0])\n",
    "print(\"Number of samples with confidently imputed MHC I:\", df_clean_hla1.shape[0])\n",
    "print(\"Number of samples with confidently imputed MHC II:\", df_clean_hla2.shape[0])\n",
    "\n",
    "# identify people who have correctly genotyped MHC class I and class II \n",
    "ids_hla1 = df_clean_hla1.Person_ID.tolist()\n",
    "ids_hla2 = df_clean_hla2.Person_ID.tolist()\n",
    "ids_correct_genotype = list(set(ids_hla1).intersection(ids_hla2))\n",
    "print('Number of UKBB participants with correctly genotyped MHC class I and II:', len(ids_correct_genotype))\n",
    "\n",
    "# dataframe with poeple who have both MHC I and MHCII imputed\n",
    "df_clean_hla1_2 = df_hla1[df_hla1['Person_ID'].isin(ids_correct_genotype)]\n",
    "print('Number of UKBB participants with correctly genotyped MHC class I and II:', df_clean_hla1_2.shape[0]) # to check that the # people is the same \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Add age data "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of individuals for whom age data is available: 424089\n"
     ]
    }
   ],
   "source": [
    "# This dataframe has been cleaned up to only include healthy individuals\n",
    "# therefore, I will only include those individuals for whom I have age, as for the others, they are likely to have been removed due to a cancer diagnosis\n",
    "# NOTE: there are some individuals who have sth in the 'first diagnosed cancer' column etc.\n",
    "# these are not excluded because they are not treated with chemo (which could affect CH) and may have been diagnosed with dysplasia \n",
    "\n",
    "age_data = pd.read_csv('/Users/barbarawalkowiak/Desktop/msc_thesis/data/2024-04-24_healthy_pheno_more_cancer_cols.tsv', sep = '\\t')\n",
    "\n",
    "age_df = age_data[['ID_v0', 'Age.when.attended.assessment.centre_v0']]\n",
    "age_df.columns.values[0] = 'Person_ID'\n",
    "age_df.columns.values[1] = 'age'\n",
    "\n",
    "print('Number of individuals for whom age data is available:', age_df.shape[0])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of samples with confidently imputed MHC I, healthy: 384600\n",
      "Number of samples with confidently imputed MHC II, healthy: 327961\n"
     ]
    }
   ],
   "source": [
    "# add age data (DO NOT allow age to be missing since if it is missing, likely to be due to a cancer diganosis)\n",
    "df_clean_hla1_age = pd.merge(df_clean_hla1, age_df, on = 'Person_ID')\n",
    "df_clean_hla2_age = pd.merge(df_clean_hla2, age_df, on = 'Person_ID')\n",
    "df_clean_hla1_2_age = pd.merge(df_clean_hla1_2, age_df, on = 'Person_ID')\n",
    "\n",
    "# check the number of individuals is correct \n",
    "print(\"Number of samples with confidently imputed MHC I, healthy:\", df_clean_hla1_age.shape[0])\n",
    "print(\"Number of samples with confidently imputed MHC II, healthy:\", df_clean_hla2_age.shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of individuals with MHC I who were screened for CH: 384600\n",
      "Number of individuals with MHC II who were screened for CH: 327961\n",
      "Number of individuals with MHC I AND II who were screened for CH: 300638\n"
     ]
    }
   ],
   "source": [
    "# subset to only include people who were screened for CH\n",
    "df_clean_hla1_age_screened = df_clean_hla1_age[df_clean_hla1_age['Person_ID'].isin(indexes_list)]\n",
    "df_clean_hla2_age_screened = df_clean_hla2_age[df_clean_hla2_age['Person_ID'].isin(indexes_list)]\n",
    "df_clean_hla1_2_age_screened = df_clean_hla1_2_age[df_clean_hla1_2_age['Person_ID'].isin(indexes_list)]\n",
    "print('Number of individuals with MHC I who were screened for CH:', df_clean_hla1_age_screened.shape[0])\n",
    "print('Number of individuals with MHC II who were screened for CH:', df_clean_hla2_age_screened.shape[0])\n",
    "print('Number of individuals with MHC I AND II who were screened for CH:', df_clean_hla1_2_age_screened.shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Person_ID</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>position</th>\n",
       "      <th>end_position</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>depth</th>\n",
       "      <th>var_depth</th>\n",
       "      <th>VAF</th>\n",
       "      <th>variant_type</th>\n",
       "      <th>...</th>\n",
       "      <th>Giacomelli_LOF_LR</th>\n",
       "      <th>SGR_LR</th>\n",
       "      <th>Personal_and_Family_History_LR_Combined</th>\n",
       "      <th>Breast_Tumor_Pathology_LR</th>\n",
       "      <th>Population_Allele_Frequency_LR</th>\n",
       "      <th>BS2_LR</th>\n",
       "      <th>Comments</th>\n",
       "      <th>Splicing_prediction</th>\n",
       "      <th>ClinVar__March_2021</th>\n",
       "      <th>gene_var</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1000130</td>\n",
       "      <td>chr2</td>\n",
       "      <td>25234373</td>\n",
       "      <td>25234373.0</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>81</td>\n",
       "      <td>2</td>\n",
       "      <td>0.024691</td>\n",
       "      <td>SNV</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DNMT3A_R882H</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1000559</td>\n",
       "      <td>chr2</td>\n",
       "      <td>25234374</td>\n",
       "      <td>25234374.0</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>56</td>\n",
       "      <td>2</td>\n",
       "      <td>0.035714</td>\n",
       "      <td>SNV</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DNMT3A_R882C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1000724</td>\n",
       "      <td>chr2</td>\n",
       "      <td>25240313</td>\n",
       "      <td>25240313.0</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>98</td>\n",
       "      <td>2</td>\n",
       "      <td>0.020408</td>\n",
       "      <td>SNV</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DNMT3A_R771*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1001048</td>\n",
       "      <td>chr2</td>\n",
       "      <td>25247629</td>\n",
       "      <td>25247629.0</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>88</td>\n",
       "      <td>2</td>\n",
       "      <td>0.022727</td>\n",
       "      <td>SNV</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DNMT3A_R326G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1001150</td>\n",
       "      <td>chr2</td>\n",
       "      <td>25247647</td>\n",
       "      <td>25247647.0</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>74</td>\n",
       "      <td>2</td>\n",
       "      <td>0.027027</td>\n",
       "      <td>SNV</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DNMT3A_R320*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>207</th>\n",
       "      <td>5645571</td>\n",
       "      <td>chr17</td>\n",
       "      <td>7673802</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>84</td>\n",
       "      <td>30</td>\n",
       "      <td>0.357143</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>82.04</td>\n",
       "      <td>3237.15</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Pathogenic_Last_reviewed_Aug_28_2019</td>\n",
       "      <td>TP53_R273H</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>210</th>\n",
       "      <td>5744845</td>\n",
       "      <td>chr17</td>\n",
       "      <td>7675088</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>52</td>\n",
       "      <td>8</td>\n",
       "      <td>0.153846</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>6.20</td>\n",
       "      <td>22639.68</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Pathogenic_Last_reviewed_Jan_7_2020</td>\n",
       "      <td>TP53_R175H</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>211</th>\n",
       "      <td>5751556</td>\n",
       "      <td>chr17</td>\n",
       "      <td>7673802</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>62</td>\n",
       "      <td>8</td>\n",
       "      <td>0.129032</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>82.04</td>\n",
       "      <td>3237.15</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Pathogenic_Last_reviewed_Aug_28_2019</td>\n",
       "      <td>TP53_R273H</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>219</th>\n",
       "      <td>5859256</td>\n",
       "      <td>chr17</td>\n",
       "      <td>7673802</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>64</td>\n",
       "      <td>6</td>\n",
       "      <td>0.093750</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>82.04</td>\n",
       "      <td>3237.15</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Pathogenic_Last_reviewed_Aug_28_2019</td>\n",
       "      <td>TP53_R273H</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>222</th>\n",
       "      <td>5890641</td>\n",
       "      <td>chr17</td>\n",
       "      <td>7673802</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>54</td>\n",
       "      <td>9</td>\n",
       "      <td>0.166667</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>82.04</td>\n",
       "      <td>3237.15</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Pathogenic_Last_reviewed_Aug_28_2019</td>\n",
       "      <td>TP53_R273H</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10602 rows × 85 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Person_ID chromosome  position  end_position ref alt  depth  var_depth  \\\n",
       "0      1000130       chr2  25234373    25234373.0   C   T     81          2   \n",
       "1      1000559       chr2  25234374    25234374.0   G   A     56          2   \n",
       "2      1000724       chr2  25240313    25240313.0   G   A     98          2   \n",
       "3      1001048       chr2  25247629    25247629.0   G   C     88          2   \n",
       "4      1001150       chr2  25247647    25247647.0   G   A     74          2   \n",
       "..         ...        ...       ...           ...  ..  ..    ...        ...   \n",
       "207    5645571      chr17   7673802           NaN   C   T     84         30   \n",
       "210    5744845      chr17   7675088           NaN   C   T     52          8   \n",
       "211    5751556      chr17   7673802           NaN   C   T     62          8   \n",
       "219    5859256      chr17   7673802           NaN   C   T     64          6   \n",
       "222    5890641      chr17   7673802           NaN   C   T     54          9   \n",
       "\n",
       "          VAF variant_type  ... Giacomelli_LOF_LR    SGR_LR  \\\n",
       "0    0.024691          SNV  ...               NaN       NaN   \n",
       "1    0.035714          SNV  ...               NaN       NaN   \n",
       "2    0.020408          SNV  ...               NaN       NaN   \n",
       "3    0.022727          SNV  ...               NaN       NaN   \n",
       "4    0.027027          SNV  ...               NaN       NaN   \n",
       "..        ...          ...  ...               ...       ...   \n",
       "207  0.357143          NaN  ...             82.04   3237.15   \n",
       "210  0.153846          NaN  ...              6.20  22639.68   \n",
       "211  0.129032          NaN  ...             82.04   3237.15   \n",
       "219  0.093750          NaN  ...             82.04   3237.15   \n",
       "222  0.166667          NaN  ...             82.04   3237.15   \n",
       "\n",
       "     Personal_and_Family_History_LR_Combined Breast_Tumor_Pathology_LR  \\\n",
       "0                                        NaN                       NaN   \n",
       "1                                        NaN                       NaN   \n",
       "2                                        NaN                       NaN   \n",
       "3                                        NaN                       NaN   \n",
       "4                                        NaN                       NaN   \n",
       "..                                       ...                       ...   \n",
       "207                                      1.0                       1.0   \n",
       "210                                      1.0                       1.0   \n",
       "211                                      1.0                       1.0   \n",
       "219                                      1.0                       1.0   \n",
       "222                                      1.0                       1.0   \n",
       "\n",
       "    Population_Allele_Frequency_LR  BS2_LR  Comments Splicing_prediction  \\\n",
       "0                              NaN     NaN       NaN                 NaN   \n",
       "1                              NaN     NaN       NaN                 NaN   \n",
       "2                              NaN     NaN       NaN                 NaN   \n",
       "3                              NaN     NaN       NaN                 NaN   \n",
       "4                              NaN     NaN       NaN                 NaN   \n",
       "..                             ...     ...       ...                 ...   \n",
       "207                            1.0     1.0       NaN                 NaN   \n",
       "210                            1.0     1.0       NaN                 NaN   \n",
       "211                            1.0     1.0       NaN                 NaN   \n",
       "219                            1.0     1.0       NaN                 NaN   \n",
       "222                            1.0     1.0       NaN                 NaN   \n",
       "\n",
       "                      ClinVar__March_2021      gene_var  \n",
       "0                                     NaN  DNMT3A_R882H  \n",
       "1                                     NaN  DNMT3A_R882C  \n",
       "2                                     NaN  DNMT3A_R771*  \n",
       "3                                     NaN  DNMT3A_R326G  \n",
       "4                                     NaN  DNMT3A_R320*  \n",
       "..                                    ...           ...  \n",
       "207  Pathogenic_Last_reviewed_Aug_28_2019    TP53_R273H  \n",
       "210   Pathogenic_Last_reviewed_Jan_7_2020    TP53_R175H  \n",
       "211  Pathogenic_Last_reviewed_Aug_28_2019    TP53_R273H  \n",
       "219  Pathogenic_Last_reviewed_Aug_28_2019    TP53_R273H  \n",
       "222  Pathogenic_Last_reviewed_Aug_28_2019    TP53_R273H  \n",
       "\n",
       "[10602 rows x 85 columns]"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# format the df with common CH variants so it can be merged with other dataframes \n",
    "df_common_variants_all.rename(columns={'sample_ID': 'Person_ID'}, inplace=True) # rename column from sample to person \n",
    "df_common_variants_all.rename(columns={'Age.when.attended.assessment.centre_v0': 'age'}, inplace = True) # rename age column\n",
    "df_common_variants_all['gene_var'] = df_common_variants_all['varID'].str.replace(' ', '_') # create a column that will have the same format as predictions\n",
    "df_common_variants_all "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Add MHC-related parameters (heterozygosity status etc.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [],
   "source": [
    "# define a function to get names of columns which indicate alleles present in the sample \n",
    "def get_columns_with_values(row, subset_columns):\n",
    "    return [col for col in subset_columns if row[col] > 0]\n",
    "\n",
    "# define a function to check if someone is heterozygous for a given class of alleles \n",
    "def is_heterozygous_allele(row, col):\n",
    "\n",
    "    # column = column with list of A, B, C alleles present in the sample\n",
    "    \n",
    "    # count the number of alleles someone has \n",
    "    # count the number of commas + add 1 = number of entires\n",
    "    count = row[col].count(',') + 1  \n",
    "    \n",
    "    # determine if they are heterozygous or homozygous \n",
    "    if count == 2:\n",
    "        return True # has two alleles so heterozygous\n",
    "    elif count == 1:\n",
    "        return False # has 1 alleles so homozygous\n",
    "    else:\n",
    "        return None # we definitely should not have anything else !!! so return None and see in how many cases this happens"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Person_ID</th>\n",
       "      <th>DRB5_101</th>\n",
       "      <th>DRB5_102</th>\n",
       "      <th>DRB5_202</th>\n",
       "      <th>DRB5_9901</th>\n",
       "      <th>DRB4_101</th>\n",
       "      <th>DRB4_103</th>\n",
       "      <th>DRB4_9901</th>\n",
       "      <th>DRB3_101</th>\n",
       "      <th>DRB3_202</th>\n",
       "      <th>...</th>\n",
       "      <th>DPA1_201</th>\n",
       "      <th>DPA1_202</th>\n",
       "      <th>DPA1_301</th>\n",
       "      <th>DPA1_401</th>\n",
       "      <th>age</th>\n",
       "      <th>allele_II_DRB</th>\n",
       "      <th>allele_II_DPA</th>\n",
       "      <th>allele_II_DPB</th>\n",
       "      <th>allele_II_DQA</th>\n",
       "      <th>allele_II_DQB</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>4860169</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>46</td>\n",
       "      <td>DRB1_301, DRB1_1501</td>\n",
       "      <td>DPA1_103</td>\n",
       "      <td>DPB1_301, DPB1_401</td>\n",
       "      <td>DQA1_102, DQA1_501</td>\n",
       "      <td>DQB1_201, DQB1_602</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3381323</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>52</td>\n",
       "      <td>DRB1_101, DRB1_401</td>\n",
       "      <td>DPA1_103</td>\n",
       "      <td>DPB1_201, DPB1_401</td>\n",
       "      <td>DQA1_101, DQA1_301</td>\n",
       "      <td>DQB1_302, DQB1_501</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2805252</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>65</td>\n",
       "      <td>DRB1_401, DRB1_1501</td>\n",
       "      <td>DPA1_103, DPA1_202</td>\n",
       "      <td>DPB1_101, DPB1_401</td>\n",
       "      <td>DQA1_102, DQA1_301</td>\n",
       "      <td>DQB1_301, DQB1_602</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3318036</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>50</td>\n",
       "      <td>DRB1_701</td>\n",
       "      <td>DPA1_201</td>\n",
       "      <td>DPB1_1101</td>\n",
       "      <td>DQA1_201</td>\n",
       "      <td>DQB1_202</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4120291</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>41</td>\n",
       "      <td>DRB1_103, DRB1_701</td>\n",
       "      <td>DPA1_201</td>\n",
       "      <td>DPB1_1001, DPB1_1101</td>\n",
       "      <td>DQA1_101, DQA1_201</td>\n",
       "      <td>DQB1_202, DQB1_501</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 155 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Person_ID  DRB5_101  DRB5_102  DRB5_202  DRB5_9901  DRB4_101  DRB4_103  \\\n",
       "0    4860169       1.0       0.0       0.0        1.0       0.0       0.0   \n",
       "1    3381323       0.0       0.0       0.0        2.0       0.0       1.0   \n",
       "2    2805252       1.0       0.0       0.0        1.0       0.0       1.0   \n",
       "3    3318036       0.0       0.0       0.0        2.0       2.0       0.0   \n",
       "4    4120291       0.0       0.0       0.0        2.0       1.0       0.0   \n",
       "\n",
       "   DRB4_9901  DRB3_101  DRB3_202  ...  DPA1_201  DPA1_202  DPA1_301  DPA1_401  \\\n",
       "0        2.0       1.0       0.0  ...       0.0       0.0       0.0       0.0   \n",
       "1        1.0       0.0       0.0  ...       0.0       0.0       0.0       0.0   \n",
       "2        1.0       0.0       0.0  ...       0.0       1.0       0.0       0.0   \n",
       "3        0.0       0.0       0.0  ...       2.0       0.0       0.0       0.0   \n",
       "4        1.0       0.0       0.0  ...       2.0       0.0       0.0       0.0   \n",
       "\n",
       "   age        allele_II_DRB       allele_II_DPA         allele_II_DPB  \\\n",
       "0   46  DRB1_301, DRB1_1501            DPA1_103    DPB1_301, DPB1_401   \n",
       "1   52   DRB1_101, DRB1_401            DPA1_103    DPB1_201, DPB1_401   \n",
       "2   65  DRB1_401, DRB1_1501  DPA1_103, DPA1_202    DPB1_101, DPB1_401   \n",
       "3   50             DRB1_701            DPA1_201             DPB1_1101   \n",
       "4   41   DRB1_103, DRB1_701            DPA1_201  DPB1_1001, DPB1_1101   \n",
       "\n",
       "        allele_II_DQA       allele_II_DQB  \n",
       "0  DQA1_102, DQA1_501  DQB1_201, DQB1_602  \n",
       "1  DQA1_101, DQA1_301  DQB1_302, DQB1_501  \n",
       "2  DQA1_102, DQA1_301  DQB1_301, DQB1_602  \n",
       "3            DQA1_201            DQB1_202  \n",
       "4  DQA1_101, DQA1_201  DQB1_202, DQB1_501  \n",
       "\n",
       "[5 rows x 155 columns]"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# reminder: filtered data to use: df_clean_hla1_age_screened, df_clean_hla2_age_screened\n",
    "\n",
    "# apply function to class I alleles\n",
    "\n",
    "# filter columns with MHC class I alleles (start with A_, B_, C_)\n",
    "subset_A = [col for col in df_clean_hla1_age_screened.columns if col.startswith(('A_'))]\n",
    "subset_B = [col for col in df_clean_hla1_age_screened.columns if col.startswith(('B_'))]\n",
    "subset_C = [col for col in df_clean_hla1_age_screened.columns if col.startswith(('C_'))]\n",
    "\n",
    "subsets_1 = [subset_A, subset_B, subset_C]\n",
    "\n",
    "for subset in subsets_1:\n",
    "    \n",
    "    new_column_name = 'allele_I_' + subset[0][0] # the first letter will indicate class\n",
    "    df_clean_hla1_age_screened[new_column_name] = df_clean_hla1_age_screened.apply(get_columns_with_values, axis=1, subset_columns=subset).apply(lambda x: ', '.join(x))\n",
    "\n",
    "df_clean_hla1_age_screened.head()\n",
    "\n",
    "# apply function to class II alleles\n",
    "subset_DR = [col for col in df_clean_hla2_age_screened.columns if 'DRB1' in col] # we are only looking at DRB1, as DRB3/4/5 are mostly missing\n",
    "subset_DPA = [col for col in df_clean_hla2_age_screened.columns if 'DPA' in col]\n",
    "subset_DPB = [col for col in df_clean_hla2_age_screened.columns if 'DPB' in col]\n",
    "subset_DQA = [col for col in df_clean_hla2_age_screened.columns if 'DQA' in col]\n",
    "subset_DQB = [col for col in df_clean_hla2_age_screened.columns if 'DQB' in col]\n",
    "\n",
    "subsets_2 = [subset_DR, subset_DPA, subset_DPB, subset_DQA, subset_DQB]\n",
    "\n",
    "for subset in subsets_2:\n",
    "    \n",
    "    new_column_name = 'allele_II_' + subset[0][:3] # the first three letters indicate class\n",
    "    df_clean_hla2_age_screened[new_column_name] = df_clean_hla2_age_screened.apply(get_columns_with_values, axis=1, subset_columns=subset).apply(lambda x: ', '.join(x))\n",
    "\n",
    "df_clean_hla2_age_screened.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "# filter columns with MHC class I alleles (start with A_, B_, C_) / MHC class II alleles \n",
    "# list of columns we want to look at to determine heterozygosity status\n",
    "columns_hla1 = ['allele_I_A', 'allele_I_B', 'allele_I_C']\n",
    "columns_hla2 = ['allele_II_DRB', 'allele_II_DPA', 'allele_II_DPB', 'allele_II_DQA', 'allele_II_DQB']\n",
    "\n",
    "for col in columns_hla1:\n",
    "    new_column_name = 'het_' + col # the first three letters indicate class\n",
    "    df_clean_hla1_age_screened[new_column_name] = df_clean_hla1_age_screened.apply(is_heterozygous_allele, axis=1, col = col) # add relevant column to the dataframe\n",
    "\n",
    "for col in columns_hla2:\n",
    "    new_column_name = 'het_' + col # the first three letters indicate class\n",
    "    df_clean_hla2_age_screened[new_column_name] = df_clean_hla2_age_screened.apply(is_heterozygous_allele, axis=1, col = col) # add relevant column to the dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "subset_classI = [col for col in df_clean_hla1_age_screened.columns if col.startswith(('A_', 'B_', 'C_'))] # all class I alleles \n",
    "subset_classII = [col for col in df_clean_hla2_age_screened.columns if col.startswith(('D'))] # all class II alleles \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# add total heterozygosity \n",
    "\n",
    "# determine the number of unique alleles typed (so if someone has HLA_A0101, HLA_A0101, HLA_B0101, HLA_B0101, HLA_C0101, HLA_C0101, the count is 3)\n",
    "df_clean_hla1_age_screened['count_class_I'] = df_clean_hla1_age_screened[subset_classI].gt(0).sum(axis=1)\n",
    "df_clean_hla2_age_screened['count_class_II'] = df_clean_hla2_age_screened[subset_classII].gt(0).sum(axis=1)\n",
    "\n",
    "# this is so we get only counts for the 10 alleles we are looking at\n",
    "subset_classII_drb = [col for col in df_clean_hla2_age_screened.columns if col.startswith('DRB1') or col.startswith('DP') or col.startswith(('DQ'))] # all class II alleles \n",
    "df_clean_hla2_age_screened['count_class_II_drb'] = df_clean_hla2_age_screened[subset_classII_drb].gt(0).sum(axis=1)\n",
    "\n",
    "# determine the number of alleles typed (so if someone has HLA_A0101, HLA_A0101, HLA_B0101, HLA_B0101, HLA_C0101, HLA_C0101, the sum is 6)\n",
    "df_clean_hla1_age_screened['sum_class_I'] = df_clean_hla1_age_screened[subset_classI].sum(axis=1)\n",
    "df_clean_hla2_age_screened['sum_class_II'] = df_clean_hla2_age_screened[subset_classII].sum(axis=1)\n",
    "\n",
    "# note that some people may not have been fully genotyped > I guess we should think about what to do with these cases \n",
    "df_clean_hla1_age_screened['het_all_class_I'] = (df_clean_hla1_age_screened['count_class_I'] == 6) # we dropped people who did not have genotype data available so if they dont have 6 alleles they are homozygous by our definition\n",
    "df_clean_hla2_age_screened['het_all_class_II'] = (df_clean_hla2_age_screened['count_class_II'] == 10) \n",
    "\n",
    "# one way to check for total heterozygosity is to see whether you have 'TRUE' in all fields (het_allele A, B, C / het allele DRB, DPA, DPB, DQA, DQB)\n",
    "df_clean_hla1_age_screened['het_all_class_I_from_allele'] = (df_clean_hla1_age_screened['het_allele_I_A'] & df_clean_hla1_age_screened['het_allele_I_B'] & df_clean_hla1_age_screened['het_allele_I_C'])\n",
    "df_clean_hla2_age_screened['het_all_class_II_from_allele'] = (df_clean_hla2_age_screened['het_allele_II_DRB'] & df_clean_hla2_age_screened['het_allele_II_DPB'] & df_clean_hla2_age_screened['het_allele_II_DPA'] & df_clean_hla2_age_screened['het_allele_II_DQB'] & df_clean_hla2_age_screened['het_allele_II_DQA'])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Add CH status (genetic variant, VAF, read depth)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of individuals with MHC I data checked for CH status: 384600\n",
      "Number of individuals with MHC II data checked for CH status: 327961\n",
      "Number of CH samples with MHC I data: 9624\n",
      "Number of CH samples with MHC II data: 8212\n",
      "Number of CH individuals with MHC I data: 9309\n",
      "Number of CH individuals with MHC II data: 7932\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Person_ID</th>\n",
       "      <th>A_101</th>\n",
       "      <th>A_102</th>\n",
       "      <th>A_103</th>\n",
       "      <th>A_201</th>\n",
       "      <th>A_202</th>\n",
       "      <th>A_203</th>\n",
       "      <th>A_205</th>\n",
       "      <th>A_206</th>\n",
       "      <th>A_207</th>\n",
       "      <th>...</th>\n",
       "      <th>het_allele_I_C</th>\n",
       "      <th>count_class_I</th>\n",
       "      <th>sum_class_I</th>\n",
       "      <th>het_all_class_I</th>\n",
       "      <th>het_all_class_I_from_allele</th>\n",
       "      <th>VAF</th>\n",
       "      <th>gene_var</th>\n",
       "      <th>depth</th>\n",
       "      <th>var_depth</th>\n",
       "      <th>ch_status</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2812213</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>5</td>\n",
       "      <td>6.0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4860169</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>5</td>\n",
       "      <td>6.0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3381323</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>6</td>\n",
       "      <td>6.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2805252</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>5</td>\n",
       "      <td>6.0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1118855</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>6</td>\n",
       "      <td>6.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 231 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Person_ID  A_101  A_102  A_103  A_201  A_202  A_203  A_205  A_206  A_207  \\\n",
       "0    2812213    0.0    0.0    0.0    2.0    0.0    0.0    0.0    0.0    0.0   \n",
       "1    4860169    2.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   \n",
       "2    3381323    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   \n",
       "3    2805252    2.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   \n",
       "4    1118855    0.0    0.0    0.0    1.0    0.0    0.0    0.0    0.0    0.0   \n",
       "\n",
       "   ...  het_allele_I_C  count_class_I  sum_class_I  het_all_class_I  \\\n",
       "0  ...            True              5          6.0            False   \n",
       "1  ...            True              5          6.0            False   \n",
       "2  ...            True              6          6.0             True   \n",
       "3  ...            True              5          6.0            False   \n",
       "4  ...            True              6          6.0             True   \n",
       "\n",
       "   het_all_class_I_from_allele  VAF  gene_var  depth  var_depth  ch_status  \n",
       "0                        False  NaN       NaN    NaN        NaN          0  \n",
       "1                        False  NaN       NaN    NaN        NaN          0  \n",
       "2                         True  NaN       NaN    NaN        NaN          0  \n",
       "3                        False  NaN       NaN    NaN        NaN          0  \n",
       "4                         True  NaN       NaN    NaN        NaN          0  \n",
       "\n",
       "[5 rows x 231 columns]"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "# add CH status to the df with age + MHC genotype for everyone you screened \n",
    "df_common_variants_sub = df_common_variants_all[['Person_ID', 'VAF', 'gene_var', 'depth', 'var_depth']]\n",
    "df_hla1_with_ch = pd.merge(df_clean_hla1_age_screened, df_common_variants_sub, on = 'Person_ID', how = 'left')\n",
    "df_hla2_with_ch = pd.merge(df_clean_hla2_age_screened, df_common_variants_sub, on = 'Person_ID', how = 'left')\n",
    "\n",
    "print('Number of individuals with MHC I data checked for CH status:', len(df_hla1_with_ch.Person_ID.unique().tolist())) \n",
    "print('Number of individuals with MHC II data checked for CH status:', len(df_hla2_with_ch.Person_ID.unique().tolist()))\n",
    "\n",
    "# check number of cases of CH\n",
    "print('Number of CH samples with MHC I data:', df_hla1_with_ch[~df_hla1_with_ch['gene_var'].isna()].shape[0]) \n",
    "print('Number of CH samples with MHC II data:', df_hla2_with_ch[~df_hla2_with_ch['gene_var'].isna()].shape[0])\n",
    "\n",
    "print('Number of CH individuals with MHC I data:', len(df_hla1_with_ch[~df_hla1_with_ch['gene_var'].isna()].Person_ID.unique().tolist())) \n",
    "print('Number of CH individuals with MHC II data:', len(df_hla2_with_ch[~df_hla2_with_ch['gene_var'].isna()].Person_ID.unique().tolist()))\n",
    "\n",
    "# add CH status (overall, any CH variant or nothing)\n",
    "df_hla1_with_ch['ch_status'] = np.where(~df_hla1_with_ch['gene_var'].isna(), 1, 0)\n",
    "df_hla2_with_ch['ch_status'] = np.where(~df_hla2_with_ch['gene_var'].isna(), 1, 0)\n",
    "\n",
    "# in the .head(), all is NaN because most people do not have CH\n",
    "df_hla1_with_ch.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "ename": "",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31mThe Kernel crashed while executing code in the current cell or a previous cell. \n",
      "\u001b[1;31mPlease review the code in the cell(s) to identify a possible cause of the failure. \n",
      "\u001b[1;31mClick <a href='https://aka.ms/vscodeJupyterKernelCrash'>here</a> for more info. \n",
      "\u001b[1;31mView Jupyter <a href='command:jupyter.viewOutput'>log</a> for further details."
     ]
    }
   ],
   "source": [
    "# save dataframes to files \n",
    "df_hla1_with_ch.to_csv('/Users/barbarawalkowiak/Desktop/msc_thesis/results/dataframes/20240907_df_hla1_ch_status.csv')\n",
    "df_hla2_with_ch.to_csv('/Users/barbarawalkowiak/Desktop/msc_thesis/results/dataframes/20240907_df_hla2_ch_status.csv')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
