{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# MHC - CH variants on all data \n",
    "- Barbara Walkowiak (bw450)\n",
    "\n",
    "- 2024-07-10, modified 2024-09-07\n",
    "\n",
    "- In this script, for each individual screened for CH in the UKB I add MHC binding scores based on their MHC genotype (class II)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Set up"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Intel MKL WARNING: Support of Intel(R) Streaming SIMD Extensions 4.2 (Intel(R) SSE4.2) enabled only processors has been deprecated. Intel oneAPI Math Kernel Library 2025.0 will require Intel(R) Advanced Vector Extensions (Intel(R) AVX) instructions.\n",
      "Intel MKL WARNING: Support of Intel(R) Streaming SIMD Extensions 4.2 (Intel(R) SSE4.2) enabled only processors has been deprecated. Intel oneAPI Math Kernel Library 2025.0 will require Intel(R) Advanced Vector Extensions (Intel(R) AVX) instructions.\n"
     ]
    }
   ],
   "source": [
    "# IMPORTS\n",
    "\n",
    "import random\n",
    "import math\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.ticker as plticker\n",
    "from matplotlib.ticker import LinearLocator, FormatStrFormatter\n",
    "from matplotlib.patches import Polygon\n",
    "from matplotlib.patches import Patch\n",
    "from matplotlib.lines import Line2D\n",
    "from matplotlib import cm\n",
    "import scipy.special\n",
    "from scipy import integrate\n",
    "import scipy.integrate as it\n",
    "from scipy.interpolate import interp1d\n",
    "from scipy.stats import kde\n",
    "import copy\n",
    "import glob, os\n",
    "import re\n",
    "# from sklearn import datasets, linear_model\n",
    "import pandas as pd\n",
    "from decimal import *\n",
    "from operator import itemgetter    \n",
    "from collections import OrderedDict\n",
    "import timeit\n",
    "import time \n",
    "import csv\n",
    "import seaborn as sns \n",
    "import scipy as sp\n",
    "from sklearn.preprocessing import LabelEncoder\n",
    "from sklearn.decomposition import PCA\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from itertools import combinations\n",
    "from scipy.stats import kruskal\n",
    "from scipy.stats import mannwhitneyu\n",
    "from matplotlib.backends.backend_pdf import PdfPages\n",
    "import statsmodels.api as sm\n",
    "import plotly.express as px\n",
    "import kaleido\n",
    "from itertools import combinations # needed for DP / DQ allele combinations \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# specify font for plotting (Helvetica is what Caroline recommends :))\n",
    "plt.rcParams['font.family'] = 'Helvetica'\n",
    "\n",
    "# stop printing warnings \n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "# this is to make things editable in Illustrator\n",
    "plt.rcParams['pdf.fonttype'] = 42 \n",
    "\n",
    "# get current date \n",
    "timestr = time.strftime(\"%Y%m%d\") "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Read in dataframe with imputed MHC II genotype "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# read in the file (has imputed MHC II genotype + CH status (genetic variant carried, VAF) + age etc) \n",
    "df_hla2 = pd.read_csv('/Users/barbarawalkowiak/Desktop/msc_thesis/results/dataframes/20240907_df_hla2_ch_status.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# will need to have combinations of DP and DQ alleles: define necessary functions\n",
    "\n",
    "# I do not want to predict stuff for 2 alpha or 2 beta chains because these are not created so we can remove these columns\n",
    "def is_valid_column_name(col):\n",
    "    parts = col.split('-') # two part of the name of the column\n",
    "    return 'A' in parts[0] and 'B' in parts[1] or 'B' in parts[0] and 'A' in parts[1]\n",
    "\n",
    "# we also want to rename columns from XXB-XXA to XXA-XXB\n",
    "def switch_parts(col):\n",
    "    parts = col.split('-') \n",
    "    new_name = parts[1] + '-' + parts[0]\n",
    "    return new_name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# for DP and DQ alleles, what matters for binding is the combination of alleles\n",
    "\n",
    "# subset dataframes for specific allele classes \n",
    "df_hla2_dp = df_hla2.filter(regex='^(Person_|DP)')\n",
    "df_hla2_dq = df_hla2.filter(regex='^(Person_|DQ)')\n",
    "df_hla2_dr = df_hla2.filter(regex='^(Person_|DR)')\n",
    "\n",
    "## DP\n",
    "# combinations for DP \n",
    "cc = list(combinations(df_hla2_dp.columns[1:],2)) # possible combinations of these columns (first is Person_ID)\n",
    "df_hla2_dp_comb = pd.concat([df_hla2_dp[c[0]].multiply(df_hla2_dp[c[1]]) for c in cc], axis=1, keys=cc) \n",
    "df_hla2_dp_comb.columns = df_hla2_dp_comb.columns.map('-'.join) # join with a '-'\n",
    "\n",
    "## DQ \n",
    "# combinations for DQ\n",
    "cc = list(combinations(df_hla2_dq.columns[1:],2)) # possible combinations of these columns (first is Person_ID)\n",
    "df_hla2_dq_comb = pd.concat([df_hla2_dq[c[0]].multiply(df_hla2_dq[c[1]]) for c in cc], axis=1, keys=cc) \n",
    "df_hla2_dq_comb.columns = df_hla2_dq_comb.columns.map('-'.join) # join with a '-'\n",
    "\n",
    "# remove columns with BB or AA combinations (only retain valid column names)\n",
    "filtered_columns_dp = [col for col in df_hla2_dp_comb.columns if is_valid_column_name(col)]\n",
    "filtered_columns_dq = [col for col in df_hla2_dq_comb.columns if is_valid_column_name(col)]\n",
    "\n",
    "# Create a new df with the filtered and reordered columns\n",
    "new_df_dp = df_hla2_dp_comb[filtered_columns_dp]\n",
    "new_df_dq = df_hla2_dq_comb[filtered_columns_dq]\n",
    "\n",
    "renamed_columns_dp = [switch_parts(col) for col in new_df_dp.columns]\n",
    "renamed_columns_dq = [switch_parts(col) for col in new_df_dq.columns]\n",
    "\n",
    "new_df_dp.columns = renamed_columns_dp\n",
    "new_df_dq.columns = renamed_columns_dq\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Person_ID</th>\n",
       "      <th>DQB1_201</th>\n",
       "      <th>DQB1_202</th>\n",
       "      <th>DQB1_301</th>\n",
       "      <th>DQB1_302</th>\n",
       "      <th>DQB1_303</th>\n",
       "      <th>DQB1_304</th>\n",
       "      <th>DQB1_401</th>\n",
       "      <th>DQB1_402</th>\n",
       "      <th>DQB1_501</th>\n",
       "      <th>...</th>\n",
       "      <th>DQA1_102-DQB1_609</th>\n",
       "      <th>DQA1_103-DQB1_609</th>\n",
       "      <th>DQA1_201-DQB1_609</th>\n",
       "      <th>DQA1_301-DQB1_609</th>\n",
       "      <th>DQA1_302-DQB1_609</th>\n",
       "      <th>DQA1_303-DQB1_609</th>\n",
       "      <th>DQA1_401-DQB1_609</th>\n",
       "      <th>DQA1_501-DQB1_609</th>\n",
       "      <th>DQA1_505-DQB1_609</th>\n",
       "      <th>DQA1_601-DQB1_609</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>4860169</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3381323</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2805252</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3318036</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4120291</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>328236</th>\n",
       "      <td>4478244</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>328237</th>\n",
       "      <td>4478244</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>328238</th>\n",
       "      <td>3573995</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>328239</th>\n",
       "      <td>3025735</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>328240</th>\n",
       "      <td>1381182</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>328241 rows × 235 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Person_ID  DQB1_201  DQB1_202  DQB1_301  DQB1_302  DQB1_303  DQB1_304  \\\n",
       "0         4860169       1.0       0.0       0.0       0.0       0.0       0.0   \n",
       "1         3381323       0.0       0.0       0.0       1.0       0.0       0.0   \n",
       "2         2805252       0.0       0.0       1.0       0.0       0.0       0.0   \n",
       "3         3318036       0.0       1.0       0.0       0.0       0.0       0.0   \n",
       "4         4120291       0.0       1.0       0.0       0.0       0.0       0.0   \n",
       "...           ...       ...       ...       ...       ...       ...       ...   \n",
       "328236    4478244       0.0       0.0       1.0       0.0       0.0       0.0   \n",
       "328237    4478244       0.0       0.0       1.0       0.0       0.0       0.0   \n",
       "328238    3573995       1.0       0.0       0.0       0.0       0.0       0.0   \n",
       "328239    3025735       0.0       0.0       1.0       0.0       0.0       0.0   \n",
       "328240    1381182       1.0       0.0       0.0       0.0       0.0       0.0   \n",
       "\n",
       "        DQB1_401  DQB1_402  DQB1_501  ...  DQA1_102-DQB1_609  \\\n",
       "0            0.0       0.0       0.0  ...                0.0   \n",
       "1            0.0       0.0       1.0  ...                0.0   \n",
       "2            0.0       0.0       0.0  ...                0.0   \n",
       "3            0.0       0.0       0.0  ...                0.0   \n",
       "4            0.0       0.0       1.0  ...                0.0   \n",
       "...          ...       ...       ...  ...                ...   \n",
       "328236       0.0       0.0       0.0  ...                0.0   \n",
       "328237       0.0       0.0       0.0  ...                0.0   \n",
       "328238       0.0       0.0       0.0  ...                0.0   \n",
       "328239       0.0       1.0       0.0  ...                0.0   \n",
       "328240       0.0       0.0       0.0  ...                0.0   \n",
       "\n",
       "        DQA1_103-DQB1_609  DQA1_201-DQB1_609  DQA1_301-DQB1_609  \\\n",
       "0                     0.0                0.0                0.0   \n",
       "1                     0.0                0.0                0.0   \n",
       "2                     0.0                0.0                0.0   \n",
       "3                     0.0                0.0                0.0   \n",
       "4                     0.0                0.0                0.0   \n",
       "...                   ...                ...                ...   \n",
       "328236                0.0                0.0                0.0   \n",
       "328237                0.0                0.0                0.0   \n",
       "328238                0.0                0.0                0.0   \n",
       "328239                0.0                0.0                0.0   \n",
       "328240                0.0                0.0                0.0   \n",
       "\n",
       "        DQA1_302-DQB1_609  DQA1_303-DQB1_609  DQA1_401-DQB1_609  \\\n",
       "0                     0.0                0.0                0.0   \n",
       "1                     0.0                0.0                0.0   \n",
       "2                     0.0                0.0                0.0   \n",
       "3                     0.0                0.0                0.0   \n",
       "4                     0.0                0.0                0.0   \n",
       "...                   ...                ...                ...   \n",
       "328236                0.0                0.0                0.0   \n",
       "328237                0.0                0.0                0.0   \n",
       "328238                0.0                0.0                0.0   \n",
       "328239                0.0                0.0                0.0   \n",
       "328240                0.0                0.0                0.0   \n",
       "\n",
       "        DQA1_501-DQB1_609  DQA1_505-DQB1_609  DQA1_601-DQB1_609  \n",
       "0                     0.0                0.0                0.0  \n",
       "1                     0.0                0.0                0.0  \n",
       "2                     0.0                0.0                0.0  \n",
       "3                     0.0                0.0                0.0  \n",
       "4                     0.0                0.0                0.0  \n",
       "...                   ...                ...                ...  \n",
       "328236                0.0                0.0                0.0  \n",
       "328237                0.0                0.0                0.0  \n",
       "328238                0.0                0.0                0.0  \n",
       "328239                0.0                0.0                0.0  \n",
       "328240                0.0                0.0                0.0  \n",
       "\n",
       "[328241 rows x 235 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# combine with the rest of the columns \n",
    "\n",
    "# DP\n",
    "df_hla2_dp_all = pd.concat([df_hla2_dp, new_df_dp], axis = 1)\n",
    "df_hla2_dp_all.replace(2, 1, inplace=True) # if someone got a 2 (2 * 1), replace to 1\n",
    "df_hla2_dp_all.replace(4, 1, inplace=True) # if someone got a 2 (2 * 2), replace to 1\n",
    "\n",
    "# let's remove columns where noone has the combination (it's possible that some combinations are never seen in our dataset)\n",
    "colsums = pd.DataFrame(df_hla2_dp_all.sum()).reset_index()\n",
    "colsums.rename(columns={'index': 'col_name', 0:'col_sum'}, inplace=True)\n",
    "col_to_retain = colsums[colsums['col_sum']>=1]['col_name'] # these are the columns to retain \n",
    "df_hla2_dp_all = df_hla2_dp_all[col_to_retain]\n",
    "df_hla2_dp_all \n",
    "\n",
    "# DQ\n",
    "df_hla2_dq_all = pd.concat([df_hla2_dq, new_df_dq], axis = 1)\n",
    "df_hla2_dq_all.replace(2, 1, inplace=True) # if someone got a 2 (2 * 1), replace to 1\n",
    "df_hla2_dq_all.replace(4, 1, inplace=True) # if someone got a 2 (2 * 2), replace to 1\n",
    "colsums = pd.DataFrame(df_hla2_dq_all.sum()).reset_index()\n",
    "colsums.rename(columns={'index': 'col_name', 0:'col_sum'}, inplace=True)\n",
    "col_to_retain = colsums[colsums['col_sum']>=1]['col_name'] # these are the columns to retain \n",
    "df_hla2_dq_all = df_hla2_dq_all[col_to_retain]\n",
    "df_hla2_dq_all  \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of samples with annotated variants CH variant and HLA genotype (DR): 327961\n",
      "Number of samples with annotated variants CH variant and HLA genotype (DP): 327961\n",
      "Number of samples with annotated variants CH variant and HLA genotype (DQ): 327961\n"
     ]
    }
   ],
   "source": [
    "# now you need to merge the data (df_hla2, with alleles and CH status including genetic variant carried, VAF etc) with files with combinations\n",
    "# select combinations\n",
    "df_hla2_dp_comb = df_hla2_dp_all.filter(regex=r'^(Person_|.*-.*)') # either column starts with Person_ID or has a hyphen somewhere (indivative of combination)\n",
    "df_hla2_dq_comb = df_hla2_dq_all.filter(regex=r'^(Person_|.*-.*)') # either column starts with Person_ID or has a hyphen somewhere (indivative of combination)\n",
    "\n",
    "# add HLA genotype data to samples with annotated CH variant and age\n",
    "df_hla2_ch = df_hla2[df_hla2.columns.drop(list(df_hla2.filter(regex='^(DQ|DP|DR)')))] # remove columns with imputation specific alleles \n",
    "df_hla2_dr = pd.merge(df_hla2_ch, df_hla2_dr, on = 'Person_ID') \n",
    "df_hla2_dp = pd.merge(df_hla2_ch, df_hla2_dp_comb, on = 'Person_ID')\n",
    "df_hla2_dq = pd.merge(df_hla2_ch, df_hla2_dq_comb, on = 'Person_ID')\n",
    "\n",
    "print('Number of samples with annotated variants CH variant and HLA genotype (DR):', len(df_hla2_dr.Person_ID.unique().tolist()))\n",
    "print('Number of samples with annotated variants CH variant and HLA genotype (DP):', len(df_hla2_dp.Person_ID.unique().tolist()))\n",
    "print('Number of samples with annotated variants CH variant and HLA genotype (DQ):', len(df_hla2_dq.Person_ID.unique().tolist()))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Person_ID</th>\n",
       "      <th>DQA1_101-DQB1_201</th>\n",
       "      <th>DQA1_102-DQB1_201</th>\n",
       "      <th>DQA1_103-DQB1_201</th>\n",
       "      <th>DQA1_104-DQB1_201</th>\n",
       "      <th>DQA1_201-DQB1_201</th>\n",
       "      <th>DQA1_301-DQB1_201</th>\n",
       "      <th>DQA1_302-DQB1_201</th>\n",
       "      <th>DQA1_303-DQB1_201</th>\n",
       "      <th>DQA1_401-DQB1_201</th>\n",
       "      <th>...</th>\n",
       "      <th>DQA1_102-DQB1_609</th>\n",
       "      <th>DQA1_103-DQB1_609</th>\n",
       "      <th>DQA1_201-DQB1_609</th>\n",
       "      <th>DQA1_301-DQB1_609</th>\n",
       "      <th>DQA1_302-DQB1_609</th>\n",
       "      <th>DQA1_303-DQB1_609</th>\n",
       "      <th>DQA1_401-DQB1_609</th>\n",
       "      <th>DQA1_501-DQB1_609</th>\n",
       "      <th>DQA1_505-DQB1_609</th>\n",
       "      <th>DQA1_601-DQB1_609</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>4860169</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3381323</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2805252</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3318036</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4120291</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>328235</th>\n",
       "      <td>1331351</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>328236</th>\n",
       "      <td>4478244</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>328238</th>\n",
       "      <td>3573995</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>328239</th>\n",
       "      <td>3025735</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>328240</th>\n",
       "      <td>1381182</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>327961 rows × 203 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Person_ID  DQA1_101-DQB1_201  DQA1_102-DQB1_201  DQA1_103-DQB1_201  \\\n",
       "0         4860169                0.0                1.0                0.0   \n",
       "1         3381323                0.0                0.0                0.0   \n",
       "2         2805252                0.0                0.0                0.0   \n",
       "3         3318036                0.0                0.0                0.0   \n",
       "4         4120291                0.0                0.0                0.0   \n",
       "...           ...                ...                ...                ...   \n",
       "328235    1331351                0.0                0.0                0.0   \n",
       "328236    4478244                0.0                0.0                0.0   \n",
       "328238    3573995                0.0                0.0                0.0   \n",
       "328239    3025735                0.0                0.0                0.0   \n",
       "328240    1381182                0.0                1.0                0.0   \n",
       "\n",
       "        DQA1_104-DQB1_201  DQA1_201-DQB1_201  DQA1_301-DQB1_201  \\\n",
       "0                     0.0                0.0                0.0   \n",
       "1                     0.0                0.0                0.0   \n",
       "2                     0.0                0.0                0.0   \n",
       "3                     0.0                0.0                0.0   \n",
       "4                     0.0                0.0                0.0   \n",
       "...                   ...                ...                ...   \n",
       "328235                0.0                0.0                0.0   \n",
       "328236                0.0                0.0                0.0   \n",
       "328238                0.0                0.0                0.0   \n",
       "328239                0.0                0.0                0.0   \n",
       "328240                0.0                0.0                0.0   \n",
       "\n",
       "        DQA1_302-DQB1_201  DQA1_303-DQB1_201  DQA1_401-DQB1_201  ...  \\\n",
       "0                     0.0                0.0                0.0  ...   \n",
       "1                     0.0                0.0                0.0  ...   \n",
       "2                     0.0                0.0                0.0  ...   \n",
       "3                     0.0                0.0                0.0  ...   \n",
       "4                     0.0                0.0                0.0  ...   \n",
       "...                   ...                ...                ...  ...   \n",
       "328235                0.0                0.0                0.0  ...   \n",
       "328236                0.0                0.0                0.0  ...   \n",
       "328238                0.0                0.0                0.0  ...   \n",
       "328239                0.0                0.0                0.0  ...   \n",
       "328240                0.0                0.0                0.0  ...   \n",
       "\n",
       "        DQA1_102-DQB1_609  DQA1_103-DQB1_609  DQA1_201-DQB1_609  \\\n",
       "0                     0.0                0.0                0.0   \n",
       "1                     0.0                0.0                0.0   \n",
       "2                     0.0                0.0                0.0   \n",
       "3                     0.0                0.0                0.0   \n",
       "4                     0.0                0.0                0.0   \n",
       "...                   ...                ...                ...   \n",
       "328235                0.0                0.0                0.0   \n",
       "328236                0.0                0.0                0.0   \n",
       "328238                0.0                0.0                0.0   \n",
       "328239                0.0                0.0                0.0   \n",
       "328240                0.0                0.0                0.0   \n",
       "\n",
       "        DQA1_301-DQB1_609  DQA1_302-DQB1_609  DQA1_303-DQB1_609  \\\n",
       "0                     0.0                0.0                0.0   \n",
       "1                     0.0                0.0                0.0   \n",
       "2                     0.0                0.0                0.0   \n",
       "3                     0.0                0.0                0.0   \n",
       "4                     0.0                0.0                0.0   \n",
       "...                   ...                ...                ...   \n",
       "328235                0.0                0.0                0.0   \n",
       "328236                0.0                0.0                0.0   \n",
       "328238                0.0                0.0                0.0   \n",
       "328239                0.0                0.0                0.0   \n",
       "328240                0.0                0.0                0.0   \n",
       "\n",
       "        DQA1_401-DQB1_609  DQA1_501-DQB1_609  DQA1_505-DQB1_609  \\\n",
       "0                     0.0                0.0                0.0   \n",
       "1                     0.0                0.0                0.0   \n",
       "2                     0.0                0.0                0.0   \n",
       "3                     0.0                0.0                0.0   \n",
       "4                     0.0                0.0                0.0   \n",
       "...                   ...                ...                ...   \n",
       "328235                0.0                0.0                0.0   \n",
       "328236                0.0                0.0                0.0   \n",
       "328238                0.0                0.0                0.0   \n",
       "328239                0.0                0.0                0.0   \n",
       "328240                0.0                0.0                0.0   \n",
       "\n",
       "        DQA1_601-DQB1_609  \n",
       "0                     0.0  \n",
       "1                     0.0  \n",
       "2                     0.0  \n",
       "3                     0.0  \n",
       "4                     0.0  \n",
       "...                   ...  \n",
       "328235                0.0  \n",
       "328236                0.0  \n",
       "328238                0.0  \n",
       "328239                0.0  \n",
       "328240                0.0  \n",
       "\n",
       "[327961 rows x 203 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# remove duplicated rows \n",
    "df_hla2_dr = df_hla2_dr.drop_duplicates()\n",
    "df_hla2_dp = df_hla2_dp.drop_duplicates()\n",
    "df_hla2_dq = df_hla2_dq.drop_duplicates()\n",
    "df_hla2_dp_comb = df_hla2_dp_comb.drop_duplicates()\n",
    "df_hla2_dq_comb = df_hla2_dq_comb.drop_duplicates()\n",
    "df_hla2_dq_comb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Person_ID</th>\n",
       "      <th>age</th>\n",
       "      <th>allele_II_DRB</th>\n",
       "      <th>allele_II_DPA</th>\n",
       "      <th>allele_II_DPB</th>\n",
       "      <th>allele_II_DQA</th>\n",
       "      <th>allele_II_DQB</th>\n",
       "      <th>het_allele_II_DRB</th>\n",
       "      <th>het_allele_II_DPA</th>\n",
       "      <th>het_allele_II_DPB</th>\n",
       "      <th>...</th>\n",
       "      <th>DQA1_102-DQB1_609</th>\n",
       "      <th>DQA1_103-DQB1_609</th>\n",
       "      <th>DQA1_201-DQB1_609</th>\n",
       "      <th>DQA1_301-DQB1_609</th>\n",
       "      <th>DQA1_302-DQB1_609</th>\n",
       "      <th>DQA1_303-DQB1_609</th>\n",
       "      <th>DQA1_401-DQB1_609</th>\n",
       "      <th>DQA1_501-DQB1_609</th>\n",
       "      <th>DQA1_505-DQB1_609</th>\n",
       "      <th>DQA1_601-DQB1_609</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>4860169</td>\n",
       "      <td>46</td>\n",
       "      <td>DRB1_301, DRB1_1501</td>\n",
       "      <td>DPA1_103</td>\n",
       "      <td>DPB1_301, DPB1_401</td>\n",
       "      <td>DQA1_102, DQA1_501</td>\n",
       "      <td>DQB1_201, DQB1_602</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3381323</td>\n",
       "      <td>52</td>\n",
       "      <td>DRB1_101, DRB1_401</td>\n",
       "      <td>DPA1_103</td>\n",
       "      <td>DPB1_201, DPB1_401</td>\n",
       "      <td>DQA1_101, DQA1_301</td>\n",
       "      <td>DQB1_302, DQB1_501</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2805252</td>\n",
       "      <td>65</td>\n",
       "      <td>DRB1_401, DRB1_1501</td>\n",
       "      <td>DPA1_103, DPA1_202</td>\n",
       "      <td>DPB1_101, DPB1_401</td>\n",
       "      <td>DQA1_102, DQA1_301</td>\n",
       "      <td>DQB1_301, DQB1_602</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3318036</td>\n",
       "      <td>50</td>\n",
       "      <td>DRB1_701</td>\n",
       "      <td>DPA1_201</td>\n",
       "      <td>DPB1_1101</td>\n",
       "      <td>DQA1_201</td>\n",
       "      <td>DQB1_202</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4120291</td>\n",
       "      <td>41</td>\n",
       "      <td>DRB1_103, DRB1_701</td>\n",
       "      <td>DPA1_201</td>\n",
       "      <td>DPB1_1001, DPB1_1101</td>\n",
       "      <td>DQA1_101, DQA1_201</td>\n",
       "      <td>DQB1_202, DQB1_501</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>328236</th>\n",
       "      <td>4478244</td>\n",
       "      <td>55</td>\n",
       "      <td>DRB1_1104, DRB1_1501</td>\n",
       "      <td>DPA1_103</td>\n",
       "      <td>DPB1_301, DPB1_402</td>\n",
       "      <td>DQA1_102, DQA1_501</td>\n",
       "      <td>DQB1_301, DQB1_602</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>328237</th>\n",
       "      <td>4478244</td>\n",
       "      <td>55</td>\n",
       "      <td>DRB1_1104, DRB1_1501</td>\n",
       "      <td>DPA1_103</td>\n",
       "      <td>DPB1_301, DPB1_402</td>\n",
       "      <td>DQA1_102, DQA1_501</td>\n",
       "      <td>DQB1_301, DQB1_602</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>328238</th>\n",
       "      <td>3573995</td>\n",
       "      <td>62</td>\n",
       "      <td>DRB1_301</td>\n",
       "      <td>DPA1_103</td>\n",
       "      <td>DPB1_401</td>\n",
       "      <td>DQA1_501</td>\n",
       "      <td>DQB1_201</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>328239</th>\n",
       "      <td>3025735</td>\n",
       "      <td>51</td>\n",
       "      <td>DRB1_404, DRB1_1101</td>\n",
       "      <td>DPA1_103</td>\n",
       "      <td>DPB1_201, DPB1_401</td>\n",
       "      <td>DQA1_301, DQA1_501</td>\n",
       "      <td>DQB1_301, DQB1_402</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>328240</th>\n",
       "      <td>1381182</td>\n",
       "      <td>67</td>\n",
       "      <td>DRB1_301, DRB1_1501</td>\n",
       "      <td>DPA1_103, DPA1_201</td>\n",
       "      <td>DPB1_101, DPB1_401</td>\n",
       "      <td>DQA1_102, DQA1_501</td>\n",
       "      <td>DQB1_201, DQB1_502</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>328241 rows × 468 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Person_ID  age         allele_II_DRB       allele_II_DPA  \\\n",
       "0         4860169   46   DRB1_301, DRB1_1501            DPA1_103   \n",
       "1         3381323   52    DRB1_101, DRB1_401            DPA1_103   \n",
       "2         2805252   65   DRB1_401, DRB1_1501  DPA1_103, DPA1_202   \n",
       "3         3318036   50              DRB1_701            DPA1_201   \n",
       "4         4120291   41    DRB1_103, DRB1_701            DPA1_201   \n",
       "...           ...  ...                   ...                 ...   \n",
       "328236    4478244   55  DRB1_1104, DRB1_1501            DPA1_103   \n",
       "328237    4478244   55  DRB1_1104, DRB1_1501            DPA1_103   \n",
       "328238    3573995   62              DRB1_301            DPA1_103   \n",
       "328239    3025735   51   DRB1_404, DRB1_1101            DPA1_103   \n",
       "328240    1381182   67   DRB1_301, DRB1_1501  DPA1_103, DPA1_201   \n",
       "\n",
       "               allele_II_DPB       allele_II_DQA       allele_II_DQB  \\\n",
       "0         DPB1_301, DPB1_401  DQA1_102, DQA1_501  DQB1_201, DQB1_602   \n",
       "1         DPB1_201, DPB1_401  DQA1_101, DQA1_301  DQB1_302, DQB1_501   \n",
       "2         DPB1_101, DPB1_401  DQA1_102, DQA1_301  DQB1_301, DQB1_602   \n",
       "3                  DPB1_1101            DQA1_201            DQB1_202   \n",
       "4       DPB1_1001, DPB1_1101  DQA1_101, DQA1_201  DQB1_202, DQB1_501   \n",
       "...                      ...                 ...                 ...   \n",
       "328236    DPB1_301, DPB1_402  DQA1_102, DQA1_501  DQB1_301, DQB1_602   \n",
       "328237    DPB1_301, DPB1_402  DQA1_102, DQA1_501  DQB1_301, DQB1_602   \n",
       "328238              DPB1_401            DQA1_501            DQB1_201   \n",
       "328239    DPB1_201, DPB1_401  DQA1_301, DQA1_501  DQB1_301, DQB1_402   \n",
       "328240    DPB1_101, DPB1_401  DQA1_102, DQA1_501  DQB1_201, DQB1_502   \n",
       "\n",
       "        het_allele_II_DRB  het_allele_II_DPA  het_allele_II_DPB  ...  \\\n",
       "0                    True              False               True  ...   \n",
       "1                    True              False               True  ...   \n",
       "2                    True               True               True  ...   \n",
       "3                   False              False              False  ...   \n",
       "4                    True              False               True  ...   \n",
       "...                   ...                ...                ...  ...   \n",
       "328236               True              False               True  ...   \n",
       "328237               True              False               True  ...   \n",
       "328238              False              False              False  ...   \n",
       "328239               True              False               True  ...   \n",
       "328240               True               True               True  ...   \n",
       "\n",
       "        DQA1_102-DQB1_609  DQA1_103-DQB1_609  DQA1_201-DQB1_609  \\\n",
       "0                     0.0                0.0                0.0   \n",
       "1                     0.0                0.0                0.0   \n",
       "2                     0.0                0.0                0.0   \n",
       "3                     0.0                0.0                0.0   \n",
       "4                     0.0                0.0                0.0   \n",
       "...                   ...                ...                ...   \n",
       "328236                0.0                0.0                0.0   \n",
       "328237                0.0                0.0                0.0   \n",
       "328238                0.0                0.0                0.0   \n",
       "328239                0.0                0.0                0.0   \n",
       "328240                0.0                0.0                0.0   \n",
       "\n",
       "        DQA1_301-DQB1_609  DQA1_302-DQB1_609  DQA1_303-DQB1_609  \\\n",
       "0                     0.0                0.0                0.0   \n",
       "1                     0.0                0.0                0.0   \n",
       "2                     0.0                0.0                0.0   \n",
       "3                     0.0                0.0                0.0   \n",
       "4                     0.0                0.0                0.0   \n",
       "...                   ...                ...                ...   \n",
       "328236                0.0                0.0                0.0   \n",
       "328237                0.0                0.0                0.0   \n",
       "328238                0.0                0.0                0.0   \n",
       "328239                0.0                0.0                0.0   \n",
       "328240                0.0                0.0                0.0   \n",
       "\n",
       "        DQA1_401-DQB1_609  DQA1_501-DQB1_609 DQA1_505-DQB1_609  \\\n",
       "0                     0.0                0.0               0.0   \n",
       "1                     0.0                0.0               0.0   \n",
       "2                     0.0                0.0               0.0   \n",
       "3                     0.0                0.0               0.0   \n",
       "4                     0.0                0.0               0.0   \n",
       "...                   ...                ...               ...   \n",
       "328236                0.0                0.0               0.0   \n",
       "328237                0.0                0.0               0.0   \n",
       "328238                0.0                0.0               0.0   \n",
       "328239                0.0                0.0               0.0   \n",
       "328240                0.0                0.0               0.0   \n",
       "\n",
       "        DQA1_601-DQB1_609  \n",
       "0                     0.0  \n",
       "1                     0.0  \n",
       "2                     0.0  \n",
       "3                     0.0  \n",
       "4                     0.0  \n",
       "...                   ...  \n",
       "328236                0.0  \n",
       "328237                0.0  \n",
       "328238                0.0  \n",
       "328239                0.0  \n",
       "328240                0.0  \n",
       "\n",
       "[328241 rows x 468 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# create a df with all relevant columns (all HLAs together)\n",
    "df_hla2_drp = pd.merge(df_hla2_dr, df_hla2_dp_comb, on = 'Person_ID')\n",
    "df_hla2_drpq = pd.merge(df_hla2_drp, df_hla2_dq_comb, on = 'Person_ID')\n",
    "df_hla2_drpq = df_hla2_drpq.drop(columns = 'Unnamed: 0')\n",
    "df_hla2_drpq"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load binding predictions for MHC II from NetMHC II"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# we need to move the format of the DP/DQ predictions to something that matches our UKBB genotyping and everything else \n",
    "def transform_format_DPQ(input_string):\n",
    "    # Define a regular expression pattern to match the input format\n",
    "    pattern = re.compile(r'HLA-(\\w{3})(\\d{1})(\\d{4})-(\\w{3})(\\d{1})(\\d{4})') # okay so this is the pattern we are trying to match\n",
    "\n",
    "    # check if there is a match\n",
    "    match = pattern.match(input_string)\n",
    "\n",
    "    # if match, apply transformation\n",
    "    if match:\n",
    "        group1 = match.group(1) # we are not including HLA annotations, this is the name of the first allele   \n",
    "        group2 = int(match.group(2)) # there will be no zeroes, leave as it is \n",
    "        group3 = int(match.group(3)) # remove zeros at the start \n",
    "        group4 = match.group(4) # name of the second allele in the combination\n",
    "        group5 = int(match.group(5)) # there will be no zeroes, leave as it is \n",
    "        group6 = int(match.group(6)) # remove zeros at the start \n",
    "\n",
    "        # Format the output string\n",
    "        output_string = f'{group1}{group2}_{group3}-{group4}{group5}_{group6}' # stitch back \n",
    "\n",
    "        return output_string # return transformed string \n",
    "\n",
    "    # if no much, return original string \n",
    "    return 0\n",
    "\n",
    "def transform_format_DR(input_string):\n",
    "    # Define a regular expression pattern to match the input format\n",
    "    pattern = re.compile(r'(\\w{3})(\\d{1})_(\\d+)') # okay so this is the pattern we are trying to match\n",
    "\n",
    "    # check if there is a match\n",
    "    match = pattern.match(input_string)\n",
    "\n",
    "    # if match, apply transformation\n",
    "    if match:\n",
    "        group1 = match.group(1) # we are not including HLA annotations, this is the name of the first allele   \n",
    "        group2 = int(match.group(2)) # there will be no zeroes, leave as it is \n",
    "        group3 = int(match.group(3)) # remove zeros at the start \n",
    "\n",
    "        # Format the output string\n",
    "        output_string = f'{group1}{group2}_{group3}' # stitch back \n",
    "\n",
    "        return output_string # return transformed string \n",
    "\n",
    "    # if no much, return original string \n",
    "    return 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Import NetMHC II scores (has scores for all variants)\n",
    "pred_file_dr = '/Users/barbarawalkowiak/Desktop/msc_thesis/netMHCII_out/scores/20240816_NetMHC_HLA_UKBB_with_affinities_DR_bestscores_allvariants.csv' # EL scores, BA scores, EL rank, BA rank, affinity prediction\n",
    "pred_file_dpq = '/Users/barbarawalkowiak/Desktop/msc_thesis/netMHCII_out/scores/20240816_NetMHC_HLA_UKBB_with_affinities_DPQ_bestscores_allvariants.csv' # EL scores, BA scores, EL rank, BA rank, affinity prediction\n",
    "\n",
    "pred_filename_dr = pred_file_dr.split('/')[2].split('.')[0] # identify file name \n",
    "pred_filename_dpq = pred_file_dpq.split('/')[2].split('.')[0] # identify file name \n",
    "pred_method_dr = pred_file_dr.split('_out')[0] # identify method used to obtain predictions\n",
    "\n",
    "# load the csvs (have separate DR and DPQ)\n",
    "pred_df_dr = pd.read_csv(pred_file_dr) # load the csv \n",
    "pred_df_dr = pred_df_dr.rename(columns={'Affinity (nM)': 'Aff_nM'})\n",
    "pred_df_dpq = pd.read_csv(pred_file_dpq) # load the csv \n",
    "pred_df_dpq = pred_df_dpq.rename(columns={'Affinity (nM)': 'Aff_nM'})\n",
    "\n",
    "# specify HLA allele naming format \n",
    "pred_df_dpq['HLA_formatted'] = pred_df_dpq['HLA'].map(transform_format_DPQ)\n",
    "pred_df_dr['HLA_formatted'] = pred_df_dr['HLA'].map(transform_format_DR)\n",
    "\n",
    "# combine all predictions\n",
    "pred_df_all = pd.concat([pred_df_dpq, pred_df_dr], axis = 0)\n",
    "pred_df_all.head()\n",
    " \n",
    "# replace variants with STOP codons (STOP > *)\n",
    "pred_df_all['gene_var'] = pred_df_all['gene_var'].str.replace('STOP', '*')\n",
    "pred_df_all = pred_df_all.drop(['Unnamed: 0.1', 'Unnamed: 0'], axis = 1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of alleles (combinations) for which predictions are available (NetMHC II): 593\n",
      "Number of HLA alleles which have been identified in the UK BioBank: 446\n",
      "Number of unique variants with predictions available (NetMHC II): 54\n"
     ]
    }
   ],
   "source": [
    "# select required columns and sort values \n",
    "pred_df_all = pred_df_all[['HLA_formatted', 'Peptide', '%Rank_EL', 'Score_EL', '%Rank_BA', 'Score_BA', 'Aff_nM', 'gene_var', 'genotype']]\n",
    "pred_df_all = pred_df_all.sort_values(by=['HLA_formatted', 'gene_var', 'genotype'])\n",
    "\n",
    "# identify gene variants \n",
    "pred_df_all['gene_var_gt'] = pred_df_all['gene_var'] + '_' + pred_df_all['genotype'] # add complete genotype data\n",
    "pred_df_all['varID'] = pred_df_all['gene_var'] # this is a column where the variant ID is in the same format as in the CH cases dataframe (in case I need to use different ones)\n",
    "scores_pred_df_all = pred_df_all[['HLA_formatted', 'Score_EL', '%Rank_EL', 'Score_BA', '%Rank_BA', 'Aff_nM', 'varID', 'gene_var', 'gene_var_gt']] # select columns of interest\n",
    "\n",
    "# Print the results \n",
    "print('Number of alleles (combinations) for which predictions are available (NetMHC II):', len(pred_df_all.HLA_formatted.unique()))\n",
    "\n",
    "# Find MHC alleles which have been typed in the UKBB\n",
    "hla_ukbb = df_hla2_drpq.filter(regex='\\d').columns # HLA from all UKBB\n",
    "print('Number of HLA alleles which have been identified in the UK BioBank:', len(hla_ukbb))\n",
    "\n",
    "# Now look at variants\n",
    "print('Number of unique variants with predictions available (NetMHC II):',  len(pred_df_all.gene_var.unique()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'find_best_score_for_variant_carried' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[16], line 26\u001b[0m\n\u001b[1;32m     24\u001b[0m ch_hla_sub_dr \u001b[38;5;241m=\u001b[39m df_hla2_dr[hla_intersect_dr_list \u001b[38;5;241m+\u001b[39m df_hla2_dr\u001b[38;5;241m.\u001b[39mcolumns[df_hla2_dr\u001b[38;5;241m.\u001b[39mcolumns\u001b[38;5;241m.\u001b[39mstr\u001b[38;5;241m.\u001b[39mcontains(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mgene_var\u001b[39m\u001b[38;5;124m'\u001b[39m)]\u001b[38;5;241m.\u001b[39mtolist()]\n\u001b[1;32m     25\u001b[0m ch_hla_sub_dr \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mconcat([df_hla2_dr[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mPerson_ID\u001b[39m\u001b[38;5;124m\"\u001b[39m], ch_hla_sub_dr], axis\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m1\u001b[39m) \u001b[38;5;66;03m# add CH cases \u001b[39;00m\n\u001b[0;32m---> 26\u001b[0m ch_hla_sub_dr[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mscore\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m=\u001b[39m ch_hla_sub_dr\u001b[38;5;241m.\u001b[39mapply(\u001b[43mfind_best_score_for_variant_carried\u001b[49m, df\u001b[38;5;241m=\u001b[39mpred_sub_dr, param\u001b[38;5;241m=\u001b[39mparam, axis\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m1\u001b[39m) \u001b[38;5;66;03m# add score for the parameter\u001b[39;00m\n\u001b[1;32m     27\u001b[0m ch_hla_scores_dr \u001b[38;5;241m=\u001b[39m ch_hla_sub_dr\u001b[38;5;241m.\u001b[39mdropna() \u001b[38;5;66;03m# remove NA (incorrectly annotated cases)\u001b[39;00m\n\u001b[1;32m     29\u001b[0m \u001b[38;5;66;03m# merge scores with VAF and age\u001b[39;00m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'find_best_score_for_variant_carried' is not defined"
     ]
    }
   ],
   "source": [
    "# # specify parameter to obtain predictions for \n",
    "# param = '%Rank_EL'\n",
    "\n",
    "# # select df with DR / DPQ alleles\n",
    "# pred_df_dr = pred_df_all[pred_df_all['HLA_formatted'].str.contains('DR')]\n",
    "# pred_df_dpq = pred_df_all[(pred_df_all['HLA_formatted'].str.contains('DP'))|(pred_df_all['HLA_formatted'].str.contains('DQ'))]\n",
    "\n",
    "# # DR\n",
    "# pred_sub_dr = pred_df_dr[['HLA_formatted', 'gene_var_gt', param]]\n",
    "# pred_sub_dr_wide = pd.pivot(pred_sub_dr, index='gene_var_gt', columns='HLA_formatted', values=param)\n",
    "# pred_sub_dr_wide = pred_sub_dr_wide.reset_index() # this is to make sure that you have the gene_var column in there too\n",
    "\n",
    "# hla_ukbb_dr = df_hla2_dr.filter(regex='\\d').columns.tolist() # relevant HLAs\n",
    "# hla_intersect_dr = pred_sub_dr_wide.columns[pred_sub_dr_wide.columns.isin(hla_ukbb_dr)] # HLA in the UKBB which I have predictions for \n",
    "# hla_intersect_dr_list = hla_intersect_dr.tolist() \n",
    "\n",
    "# # prepare gene variants names to match names in the Patient file \n",
    "# # pred_sub_dr = pred_sub_dr_wide[hla_intersect_dr_list + pred_sub_dr_wide.columns[pred_sub_dr_wide.columns.str.contains('gene_var')].tolist()]\n",
    "# pred_sub_dr = pred_sub_dr[pred_sub_dr['gene_var_gt'].str.contains('_ch', regex=True)] # retain CH scores only \n",
    "# pred_sub_dr['gene_var'] = pred_sub_dr['gene_var_gt'].str.replace('_ch', '') # remove the ch / refseq annotation\n",
    "# pred_sub_dr['gene_var'] = pred_sub_dr['gene_var'].str.replace('_refseq', '') # remove refseq if present \n",
    "# pred_sub_dr['gene_var'] = pred_sub_dr['gene_var'].astype(str)\n",
    "\n",
    "# ch_hla_sub_dr = df_hla2_dr[hla_intersect_dr_list + df_hla2_dr.columns[df_hla2_dr.columns.str.contains('gene_var')].tolist()]\n",
    "# ch_hla_sub_dr = pd.concat([df_hla2_dr[\"Person_ID\"], ch_hla_sub_dr], axis=1) # add CH cases \n",
    "# ch_hla_sub_dr['score'] = ch_hla_sub_dr.apply(find_best_score_for_variant_carried, df=pred_sub_dr, param=param, axis=1) # add score for the parameter\n",
    "# ch_hla_scores_dr = ch_hla_sub_dr.dropna() # remove NA (incorrectly annotated cases)\n",
    "\n",
    "# # merge scores with VAF and age\n",
    "# age_vaf_dr = df_hla2_dr[['Person_ID', 'VAF', 'var_depth', 'depth', 'age', 'gene_var']]\n",
    "# ch_hla_merge_dr = pd.merge(ch_hla_scores_dr, age_vaf_dr, on = ['Person_ID', 'gene_var'])\n",
    "\n",
    "# # now add the columns with VAF and age \n",
    "# col_to_select = ['Person_ID', 'gene_var', 'score', 'age', 'var_depth', 'VAF'] # subset the data \n",
    "# ch_hla_merge_sub_dr = ch_hla_merge_dr[col_to_select]\n",
    "# ch_hla_merge_sub_dr['log_score'] = -1*np.log10(ch_hla_merge_sub_dr['score']) # convert score to -log10(score)\n",
    "# ch_hla_merge_sub_dr['allele_type'] = 'DR'\n",
    "\n",
    "# # DP\n",
    "# pred_sub_dpq = pred_df_dpq[['HLA_formatted', 'gene_var_gt', param]]\n",
    "# pred_sub_dpq = pred_sub_dpq[~pred_sub_dpq.duplicated()] # remove duplicate columns \n",
    "# pred_sub_dpq_reset = pred_sub_dpq.reset_index()\n",
    "# pred_sub_dpq_wide = pd.pivot(pred_sub_dpq_reset, index='gene_var_gt', columns='HLA_formatted', values=param)\n",
    "# pred_sub_dpq_wide = pred_sub_dpq_wide.reset_index() # this is to make sure that you have the gene_var column in there too\n",
    "\n",
    "# hla_ukbb_dp = df_hla2_dp.filter(regex='\\d').columns.tolist() # relevant HLAs\n",
    "# hla_intersect_dp = pred_sub_dpq_wide.columns[pred_sub_dpq_wide.columns.isin(hla_ukbb_dp)] # HLA in the UKBB which I have predictions for \n",
    "# hla_intersect_dp_list = hla_intersect_dp.tolist() \n",
    "\n",
    "# # prepare gene variants names to match names in the Patient file \n",
    "# pred_sub_dp = pred_sub_dpq_wide[hla_intersect_dp_list + pred_sub_dpq_wide.columns[pred_sub_dpq_wide.columns.str.contains('gene_var')].tolist()]\n",
    "# pred_sub_dp = pred_sub_dp[pred_sub_dp['gene_var_gt'].str.contains('_ch', regex=True)] # retain CH scores only \n",
    "# pred_sub_dp['gene_var'] = pred_sub_dp['gene_var_gt'].str.replace('_ch', '') # remove the ch / refseq annotation\n",
    "# pred_sub_dp['gene_var'] = pred_sub_dp['gene_var'].str.replace('_refseq', '') # remove refseq if present \n",
    "# pred_sub_dp['gene_var'] = pred_sub_dp['gene_var'].astype(str)\n",
    "\n",
    "# # look at predictions for a specific allele class  \n",
    "# ch_hla_sub_dp = df_hla2_dp[hla_intersect_dp_list + df_hla2_dp.columns[df_hla2_dp.columns.str.contains('gene_var')].tolist()]\n",
    "# ch_hla_sub_dp = pd.concat([df_hla2_dp[\"Person_ID\"], ch_hla_sub_dp], axis=1) # add CH cases \n",
    "# ch_hla_sub_dp['score'] = ch_hla_sub_dp.apply(find_best_score_for_variant_carried, df=pred_sub_dp, param=param, axis=1) # add score for the parameter\n",
    "# ch_hla_scores_dp = ch_hla_sub_dp.dropna() # remove NA (incorrectly annotated cases)\n",
    "\n",
    "# # merge scores with VAF and age\n",
    "# age_vaf_dp = df_hla2_dp[['Person_ID', 'VAF', 'var_depth', 'age', 'gene_var']]\n",
    "# ch_hla_merge_dp = pd.merge(ch_hla_scores_dp, age_vaf_dp, on = ['Person_ID', 'gene_var'])\n",
    "\n",
    "# # now add the columns with VAF and age \n",
    "# col_to_select = ['Person_ID', 'gene_var', 'score', 'age', 'var_depth', 'VAF'] # subset the data \n",
    "# ch_hla_merge_sub_dp = ch_hla_merge_dp[col_to_select]\n",
    "# ch_hla_merge_sub_dp['log_score'] = -1*np.log10(ch_hla_merge_sub_dp['score']) # convert score to -log10(score)\n",
    "# ch_hla_merge_sub_dp['allele_type'] = 'DP'\n",
    "\n",
    "# # DQ\n",
    "# hla_ukbb_dq = df_hla2_dq.filter(regex='\\d').columns.tolist() # relevant HLAs\n",
    "# hla_intersect_dq = pred_sub_dpq_wide.columns[pred_sub_dpq_wide.columns.isin(hla_ukbb_dq)] # HLA in the UKBB which I have predictions for \n",
    "# hla_intersect_dq_list = hla_intersect_dq.tolist() \n",
    "\n",
    "# # prepare gene variants names to match names in the Patient file \n",
    "# pred_sub_dq = pred_sub_dpq_wide[hla_intersect_dq_list + pred_sub_dpq_wide.columns[pred_sub_dpq_wide.columns.str.contains('gene_var')].tolist()]\n",
    "# pred_sub_dq = pred_sub_dq[pred_sub_dq['gene_var_gt'].str.contains('_ch', regex=True)] # retain CH scores only \n",
    "# pred_sub_dq['gene_var'] = pred_sub_dq['gene_var_gt'].str.replace('_ch', '') # remove the ch / refseq annotation\n",
    "# pred_sub_dq['gene_var'] = pred_sub_dq['gene_var'].str.replace('_refseq', '') # remove refseq if present \n",
    "# pred_sub_dq['gene_var'] = pred_sub_dq['gene_var'].astype(str)\n",
    "\n",
    "# # subset batch_gene_age_hla file \n",
    "# ch_hla_sub_dq = df_hla2_dq[hla_intersect_dq_list + df_hla2_dq.columns[df_hla2_dq.columns.str.contains('gene_var')].tolist()]\n",
    "# ch_hla_sub_dq = pd.concat([df_hla2_dq[\"Person_ID\"], ch_hla_sub_dq], axis=1) # add CH cases \n",
    "# ch_hla_sub_dq['score'] = ch_hla_sub_dq.apply(find_best_score_for_variant_carried, df=pred_sub_dq, param=param, axis=1) # add score for the parameter\n",
    "# ch_hla_scores_dq = ch_hla_sub_dq.dropna() # remove NA (incorrectly annotated cases)\n",
    "\n",
    "# # merge scores with VAF and age\n",
    "# age_vaf_dq = df_hla2_dq[['Person_ID', 'VAF', 'var_depth', 'age', 'gene_var']]\n",
    "# ch_hla_merge_dq = pd.merge(ch_hla_scores_dq, age_vaf_dq, on = ['Person_ID', 'gene_var'])\n",
    "\n",
    "# # now add the columns with VAF and age \n",
    "# col_to_select = ['Person_ID', 'gene_var', 'score', 'age', 'var_depth', 'VAF'] # subset the data \n",
    "# ch_hla_merge_sub_dq = ch_hla_merge_dq[col_to_select]\n",
    "# ch_hla_merge_sub_dq['log_score'] = -1*np.log10(ch_hla_merge_sub_dq['score']) # convert score to -log10(score)\n",
    "# ch_hla_merge_sub_dq['allele_type'] = 'DQ'\n",
    "\n",
    "# # we can actually combine all these scores now \n",
    "# ch_hla_merge_sub_all = pd.concat([ch_hla_merge_sub_dr, ch_hla_merge_sub_dp, ch_hla_merge_sub_dq], axis = 0)\n",
    "\n",
    "# # filter so this only includes variants with > 15 samples present\n",
    "# gene_counts = ch_hla_merge_sub_all.gene_var.value_counts().reset_index()\n",
    "# variants_to_examine = gene_counts[gene_counts['count'] > 15].gene_var.tolist()\n",
    "\n",
    "# ch_hla_merge_sub_all = ch_hla_merge_sub_all[ch_hla_merge_sub_all['gene_var'].isin(variants_to_examine)]\n",
    "# ch_hla_merge_sub_dr = ch_hla_merge_sub_dp[ch_hla_merge_sub_dp['gene_var'].isin(variants_to_examine)]\n",
    "# ch_hla_merge_sub_dp = ch_hla_merge_sub_dq[ch_hla_merge_sub_dq['gene_var'].isin(variants_to_examine)]\n",
    "# ch_hla_merge_sub_dq = ch_hla_merge_sub_dr[ch_hla_merge_sub_dr['gene_var'].isin(variants_to_examine)]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# # define function to find the best score for the variant that is carried (ie present in the person)\n",
    "# def find_best_score_for_variant_carried_all_hla(row, df, param):\n",
    "\n",
    "#     '''\n",
    "#     This functiion is applied to the CH dataset\n",
    "#     The df is the dataset with predictions for given allele and genetic variant \n",
    "#     parameter is what to base this prediction on (here will only be using %Rank_EL)\n",
    "#     '''\n",
    "    \n",
    "#     # these are the values of rows to seach for \n",
    "#     row_values = pd.to_numeric(row[1:-1], errors='coerce')\n",
    "    \n",
    "#     # find alleles which are present (NOTE this is changed from the previous function bc formatting is different)\n",
    "#     hla = row.index[1:-1][row_values >= 1]  \n",
    "    \n",
    "#     # find values corresponding to these alleles \n",
    "#     vals = df.loc[df['gene_var'] == row['gene_var'], hla].values.flatten() \n",
    "    \n",
    "#     # it can be that nothing was found e.g., bc there were no predictions made for this variant\n",
    "#     if vals.size == 0: \n",
    "#         value = None\n",
    "    \n",
    "#     else: \n",
    "    \n",
    "#         # note here we will only be using '%Rank_EL'\n",
    "#         if param == \"Aff_nM\":\n",
    "#             value = min(vals) # highest affinity corresponds to lowest value (see plots above)\n",
    "#         elif param == \"%Rank_EL\":\n",
    "#             value = min(vals) # the best rank is the lowest one (indicates peptide in top x % of binders)\n",
    "#         elif param == \"Score_EL\":\n",
    "#             value = max(vals) # the best score is the highest one \n",
    "#         elif param == \"%Rank_BA\":\n",
    "#             value = min(vals) # the best rank is the lowest one (indicates peptide in top x % of binders)\n",
    "#         elif param == \"Score_BA\":\n",
    "#             value = max(vals) # the best score is the highest one\n",
    "#         else:\n",
    "#             print('Incorrect parameter provided') \n",
    "    \n",
    "#     return value  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "ename": "",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31mThe Kernel crashed while executing code in the current cell or a previous cell. \n",
      "\u001b[1;31mPlease review the code in the cell(s) to identify a possible cause of the failure. \n",
      "\u001b[1;31mClick <a href='https://aka.ms/vscodeJupyterKernelCrash'>here</a> for more info. \n",
      "\u001b[1;31mView Jupyter <a href='command:jupyter.viewOutput'>log</a> for further details."
     ]
    }
   ],
   "source": [
    "# # specify parameter to obtain predictions for \n",
    "# param = '%Rank_EL'\n",
    "\n",
    "# # add a single score based on all of the MHC class II alleles someone has (not breaking down into DR, DP, DQ)\n",
    "# pred_sub_all = pred_df_all[['HLA_formatted', 'gene_var_gt', param]]\n",
    "# pred_sub_all = pred_sub_all[~pred_sub_all.duplicated()] # remove duplicate columns \n",
    "# pred_sub_all_reset = pred_sub_all.reset_index()\n",
    "# pred_sub_all_wide = pd.pivot(pred_sub_all_reset, index='gene_var_gt', columns='HLA_formatted', values=param).reset_index()\n",
    "\n",
    "# hla_ukbb_all = df_hla2_drpq.columns[22:].tolist() # this includes both separate DP / DQ and in combinations\n",
    "# hla_intersect_all = pred_sub_all_wide.columns[pred_sub_all_wide.columns.isin(hla_ukbb_all)] # HLA in the UKBB which I have predictions for \n",
    "# hla_intersect_all_list = hla_intersect_all.tolist() \n",
    "\n",
    "# # prepare gene variants names to match names in the Patient file \n",
    "# pred_sub_all = pred_sub_all_wide[hla_intersect_all_list + pred_sub_all_wide.columns[pred_sub_all_wide.columns.str.contains('gene_var')].tolist()]\n",
    "# pred_sub_all = pred_sub_all[pred_sub_all['gene_var_gt'].str.contains('_ch', regex=True)] # retain CH scores only \n",
    "# pred_sub_all['gene_var'] = pred_sub_all['gene_var_gt'].str.replace('_ch', '') # remove the ch / refseq annotation\n",
    "# pred_sub_all['gene_var'] = pred_sub_all['gene_var'].str.replace('_refseq', '') # remove refseq if present \n",
    "# pred_sub_all['gene_var'] = pred_sub_all['gene_var'].astype(str)\n",
    "\n",
    "# # subset batch_gene_age_hla file \n",
    "# ch_hla_sub_all = df_hla2_drpq[hla_intersect_all_list + df_hla2_drpq.columns[df_hla2_drpq.columns.str.contains('gene_var')].tolist()]\n",
    "# ch_hla_sub_all = pd.concat([df_hla2_drpq[\"Person_ID\"], ch_hla_sub_all], axis=1) # add CH cases \n",
    "# ch_hla_sub_all['score'] = ch_hla_sub_all.apply(find_best_score_for_variant_carried_all_hla, df=pred_sub_all, param=param, axis=1) # add score for the parameter\n",
    "# ch_hla_scores_all = ch_hla_sub_all.dropna() # remove NA (incorrectly annotated cases)\n",
    "\n",
    "# # merge scores with VAF and age\n",
    "# age_vaf_all = df_hla2_drpq[['Person_ID', 'VAF', 'var_depth', 'depth', 'age', 'gene_var']]\n",
    "# ch_hla_merge_all = pd.merge(ch_hla_scores_all, age_vaf_all, on = ['Person_ID', 'gene_var'])\n",
    "\n",
    "# # now add the columns with VAF and age \n",
    "# col_to_select = ['Person_ID', 'gene_var', 'score', 'age', 'var_depth', 'depth', 'VAF'] # subset the data \n",
    "# ch_hla_merge_sub_all = ch_hla_merge_all[col_to_select]\n",
    "# ch_hla_merge_sub_all['log_score'] = -1*np.log10(ch_hla_merge_sub_all['score']) # convert score to -log10(score)\n",
    "\n",
    "# # filter out rare variants\n",
    "# gene_counts = ch_hla_merge_sub_all.gene_var.value_counts().reset_index()\n",
    "# variants_to_examine = gene_counts[gene_counts['count'] > 15].gene_var.tolist()\n",
    "# ch_hla_merge_sub_all = ch_hla_merge_sub_all[ch_hla_merge_sub_all['gene_var'].isin(variants_to_examine)]\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Add scores for all variants "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# need to do something to assign, for everyone, the score for every variant we have predictions for based on their MHC I\n",
    "\n",
    "# define the function to find best scores (we can technically do this for different parameters, realistically I think will just do it for %Rank_EL)\n",
    "\n",
    "def find_best_score_for_all_variants(row, df, param):\n",
    "\n",
    "    '''\n",
    "    row = function is applied to each row of the participant dataframe (ie run through each participant)\n",
    "\n",
    "    df = dataframe with prediction scores (like netmhc)\n",
    "\n",
    "    Allowed parameters (param) are:\n",
    "    Aff_nM - affinity (raw number)\n",
    "    Score_BA - binding affinity score\n",
    "    Score_EL - elution score\n",
    "    %Rank_BA - %Rank of binding affinity cf a set of random peptides\n",
    "    %Rank_EL - %Rank of elution cf a set of random peptides\n",
    "    '''\n",
    "\n",
    "\n",
    "    # get HLAs for each person \n",
    "    hlas = row.index[7:][row[7:] >= 1] # select alleles which each Person (row) carries (the first 7 columns are: Person ID, gene_var, VAF, var_depth, depth, CH status, age)\n",
    "\n",
    "    # get variants \n",
    "    variants = df['gene_var']\n",
    "   \n",
    "    scores = {} # initialise empy dictionaries\n",
    "\n",
    "    # depending on the parameter, pick the minimum of maximum value \n",
    "    if param == \"Aff_nM\":\n",
    "        for var in variants:\n",
    "            # Find the minimum value for each variant in the category that is present\n",
    "            best_value = min(df.loc[df['gene_var'] == var, hlas].values[0])\n",
    "            # Update the dictionary with the minimum value for the corresponding variant\n",
    "            scores[f'score_{var}'] = best_value\n",
    "        return pd.Series(scores)\n",
    "\n",
    "    elif param == \"Score_BA\":\n",
    "        for var in variants:\n",
    "            \n",
    "            best_value = max(df.loc[df['gene_var'] == var, hlas].values[0])\n",
    "            scores[f'score_{var}'] = best_value\n",
    "        \n",
    "        return pd.Series(scores)\n",
    "\n",
    "    elif param == \"Score_EL\":\n",
    "        for var in variants:\n",
    "           \n",
    "            best_value = max(df.loc[df['gene_var'] == var, hlas].values[0])\n",
    "            scores[f'score_{var}'] = best_value\n",
    "\n",
    "        return pd.Series(scores)\n",
    "\n",
    "    elif param == \"%Rank_BA\":\n",
    "        for var in variants:\n",
    "\n",
    "            best_value = min(df.loc[df['gene_var'] == var, hlas].values[0])\n",
    "            scores[f'score_{var}'] = best_value\n",
    "\n",
    "        return pd.Series(scores)\n",
    "\n",
    "    # we will likely be choosing this option\n",
    "    elif param == \"%Rank_EL\":\n",
    "        for var in variants:\n",
    "            \n",
    "            best_value = min(df.loc[df['gene_var'] == var, hlas].values[0]) # choose minimum rank as the best score \n",
    "            scores[f'score_{var}'] = best_value\n",
    "\n",
    "        return pd.Series(scores)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### check all variants screened for are present "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of variants identified in the UKBB from CH screening: 40\n",
      "Number or variants I have predictions for (NetMHC): 54\n",
      "List of variants identified in UKBB: ['DNMT3A_P904L', 'DNMT3A_P904Q', 'DNMT3A_P904R', 'DNMT3A_R320*', 'DNMT3A_R326C', 'DNMT3A_R326G', 'DNMT3A_R326S', 'DNMT3A_R598*', 'DNMT3A_R729G', 'DNMT3A_R729W', 'DNMT3A_R736C', 'DNMT3A_R736G', 'DNMT3A_R736H', 'DNMT3A_R736L', 'DNMT3A_R736S', 'DNMT3A_R771*', 'DNMT3A_R882C', 'DNMT3A_R882H', 'DNMT3A_R882L', 'DNMT3A_R882P', 'DNMT3A_R882S', 'DNMT3A_Y735C', 'DNMT3A_Y735F', 'DNMT3A_Y735S', 'GNB1_K57E', 'IDH1_R132H', 'IDH2_R140Q', 'IDH2_R172K', 'JAK2_V617F', 'KRAS_G12D', 'KRAS_G12S', 'MPL_W515L', 'NRAS_G12D', 'SF3B1_K666N', 'SF3B1_K700E', 'SRSF2_P95H', 'SRSF2_P95L', 'SRSF2_P95R', 'TP53_R175H', 'TP53_R273H']\n",
      "List of variants with NetMHC preds: ['DNMT3A_P904L', 'DNMT3A_P904Q', 'DNMT3A_P904R', 'DNMT3A_R320*', 'DNMT3A_R326C', 'DNMT3A_R326G', 'DNMT3A_R326S', 'DNMT3A_R598*', 'DNMT3A_R598G', 'DNMT3A_R729G', 'DNMT3A_R729W', 'DNMT3A_R736C', 'DNMT3A_R736G', 'DNMT3A_R736H', 'DNMT3A_R736L', 'DNMT3A_R736P', 'DNMT3A_R736S', 'DNMT3A_R771*', 'DNMT3A_R771G', 'DNMT3A_R882C', 'DNMT3A_R882G', 'DNMT3A_R882H', 'DNMT3A_R882L', 'DNMT3A_R882P', 'DNMT3A_R882S', 'DNMT3A_Y735C', 'DNMT3A_Y735F', 'DNMT3A_Y735S', 'GNB1_K57E', 'IDH1_R132H', 'IDH2_R140L', 'IDH2_R140Q', 'IDH2_R172K', 'JAK2_V617F', 'KIT_D816G', 'KIT_D816V', 'KRAS_G12A', 'KRAS_G12D', 'KRAS_G12S', 'KRAS_G12V', 'MPL_W515*', 'MPL_W515L', 'NRAS_G12A', 'NRAS_G12C', 'NRAS_G12D', 'NRAS_G12S', 'NRAS_G12V', 'SF3B1_K666N', 'SF3B1_K700E', 'SRSF2_P95H', 'SRSF2_P95L', 'SRSF2_P95R', 'TP53_R175H', 'TP53_R273H']\n"
     ]
    }
   ],
   "source": [
    "print('Number of variants identified in the UKBB from CH screening:', len(df_hla2.gene_var.unique())-1) # NB this list also includes 'NaN' that's why its -1\n",
    "print('Number or variants I have predictions for (NetMHC):',  len(pred_df_all.gene_var.unique()))\n",
    "\n",
    "variants_in_ukbb = df_hla2.gene_var.unique().tolist()\n",
    "variants_in_ukbb = [x for x in variants_in_ukbb if str(x) != 'nan']\n",
    "print('List of variants identified in UKBB:', sorted(variants_in_ukbb))\n",
    "print('List of variants with NetMHC preds:',  sorted(pred_df_all.gene_var.unique())) # there is more because I also made predictions for variants which ended up not being common enough "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['DNMT3A_P904L_ch', 'DNMT3A_P904L_refseq', 'DNMT3A_P904Q_ch',\n",
       "       'DNMT3A_P904Q_refseq', 'DNMT3A_P904R_ch', 'DNMT3A_P904R_refseq',\n",
       "       'DNMT3A_R320*_ch', 'DNMT3A_R320*_refseq', 'DNMT3A_R326C_ch',\n",
       "       'DNMT3A_R326C_refseq', 'DNMT3A_R326G_ch', 'DNMT3A_R326G_refseq',\n",
       "       'DNMT3A_R326S_ch', 'DNMT3A_R326S_refseq', 'DNMT3A_R598*_ch',\n",
       "       'DNMT3A_R598*_refseq', 'DNMT3A_R598G_ch', 'DNMT3A_R598G_refseq',\n",
       "       'DNMT3A_R729G_ch', 'DNMT3A_R729G_refseq', 'DNMT3A_R729W_ch',\n",
       "       'DNMT3A_R729W_refseq', 'DNMT3A_R736C_ch', 'DNMT3A_R736C_refseq',\n",
       "       'DNMT3A_R736G_ch', 'DNMT3A_R736G_refseq', 'DNMT3A_R736H_ch',\n",
       "       'DNMT3A_R736H_refseq', 'DNMT3A_R736L_ch', 'DNMT3A_R736L_refseq',\n",
       "       'DNMT3A_R736P_ch', 'DNMT3A_R736P_refseq', 'DNMT3A_R736S_ch',\n",
       "       'DNMT3A_R736S_refseq', 'DNMT3A_R771*_ch', 'DNMT3A_R771*_refseq',\n",
       "       'DNMT3A_R771G_ch', 'DNMT3A_R771G_refseq', 'DNMT3A_R882C_ch',\n",
       "       'DNMT3A_R882C_refseq', 'DNMT3A_R882G_ch', 'DNMT3A_R882G_refseq',\n",
       "       'DNMT3A_R882H_ch', 'DNMT3A_R882H_refseq', 'DNMT3A_R882L_ch',\n",
       "       'DNMT3A_R882L_refseq', 'DNMT3A_R882P_ch', 'DNMT3A_R882P_refseq',\n",
       "       'DNMT3A_R882S_ch', 'DNMT3A_R882S_refseq', 'DNMT3A_Y735C_ch',\n",
       "       'DNMT3A_Y735C_refseq', 'DNMT3A_Y735F_ch', 'DNMT3A_Y735F_refseq',\n",
       "       'DNMT3A_Y735S_ch', 'DNMT3A_Y735S_refseq', 'GNB1_K57E_ch',\n",
       "       'GNB1_K57E_refseq', 'IDH1_R132H_ch', 'IDH1_R132H_refseq',\n",
       "       'IDH2_R140L_ch', 'IDH2_R140L_refseq', 'IDH2_R140Q_ch',\n",
       "       'IDH2_R140Q_refseq', 'IDH2_R172K_ch', 'IDH2_R172K_refseq',\n",
       "       'JAK2_V617F_ch', 'JAK2_V617F_refseq', 'KIT_D816G_ch',\n",
       "       'KIT_D816G_refseq', 'KIT_D816V_ch', 'KIT_D816V_refseq',\n",
       "       'KRAS_G12A_ch', 'KRAS_G12A_refseq', 'KRAS_G12D_ch',\n",
       "       'KRAS_G12D_refseq', 'KRAS_G12S_ch', 'KRAS_G12S_refseq',\n",
       "       'KRAS_G12V_ch', 'KRAS_G12V_refseq', 'MPL_W515*_ch',\n",
       "       'MPL_W515*_refseq', 'MPL_W515L_ch', 'MPL_W515L_refseq',\n",
       "       'NRAS_G12A_ch', 'NRAS_G12A_refseq', 'NRAS_G12C_ch',\n",
       "       'NRAS_G12C_refseq', 'NRAS_G12D_ch', 'NRAS_G12D_refseq',\n",
       "       'NRAS_G12S_ch', 'NRAS_G12S_refseq', 'NRAS_G12V_ch',\n",
       "       'NRAS_G12V_refseq', 'SF3B1_K666N_ch', 'SF3B1_K666N_refseq',\n",
       "       'SF3B1_K700E_ch', 'SF3B1_K700E_refseq', 'SRSF2_P95H_ch',\n",
       "       'SRSF2_P95H_refseq', 'SRSF2_P95L_ch', 'SRSF2_P95L_refseq',\n",
       "       'SRSF2_P95R_ch', 'SRSF2_P95R_refseq', 'TP53_R175H_ch',\n",
       "       'TP53_R175H_refseq', 'TP53_R273H_ch', 'TP53_R273H_refseq'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# make sure the naming of variants is the same\n",
    "pred_df_all['gene_var'] = pred_df_all['gene_var'].str.replace('STOP', '*')\n",
    "pred_df_all['gene_var'].unique()\n",
    "\n",
    "pred_df_all['gene_var_gt'] = pred_df_all['gene_var_gt'].str.replace('STOP', '*')\n",
    "pred_df_all['gene_var_gt'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>HLA_formatted</th>\n",
       "      <th>gene_var_gt</th>\n",
       "      <th>DPA1_103-DPB1_1001</th>\n",
       "      <th>DPA1_103-DPB1_101</th>\n",
       "      <th>DPA1_103-DPB1_1101</th>\n",
       "      <th>DPA1_103-DPB1_1301</th>\n",
       "      <th>DPA1_103-DPB1_1401</th>\n",
       "      <th>DPA1_103-DPB1_1501</th>\n",
       "      <th>DPA1_103-DPB1_1601</th>\n",
       "      <th>DPA1_103-DPB1_1701</th>\n",
       "      <th>DPA1_103-DPB1_1801</th>\n",
       "      <th>...</th>\n",
       "      <th>DRB3_101</th>\n",
       "      <th>DRB3_202</th>\n",
       "      <th>DRB3_210</th>\n",
       "      <th>DRB3_224</th>\n",
       "      <th>DRB3_301</th>\n",
       "      <th>DRB4_101</th>\n",
       "      <th>DRB4_103</th>\n",
       "      <th>DRB5_101</th>\n",
       "      <th>DRB5_102</th>\n",
       "      <th>DRB5_202</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DNMT3A_P904L_ch</td>\n",
       "      <td>20.84</td>\n",
       "      <td>7.30</td>\n",
       "      <td>6.75</td>\n",
       "      <td>5.05</td>\n",
       "      <td>36.28</td>\n",
       "      <td>21.29</td>\n",
       "      <td>8.93</td>\n",
       "      <td>17.73</td>\n",
       "      <td>17.99</td>\n",
       "      <td>...</td>\n",
       "      <td>64.62</td>\n",
       "      <td>54.75</td>\n",
       "      <td>54.75</td>\n",
       "      <td>47.57</td>\n",
       "      <td>60.38</td>\n",
       "      <td>57.86</td>\n",
       "      <td>57.86</td>\n",
       "      <td>1.54</td>\n",
       "      <td>1.54</td>\n",
       "      <td>2.05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DNMT3A_P904L_refseq</td>\n",
       "      <td>8.04</td>\n",
       "      <td>7.35</td>\n",
       "      <td>0.99</td>\n",
       "      <td>0.67</td>\n",
       "      <td>25.15</td>\n",
       "      <td>9.09</td>\n",
       "      <td>3.76</td>\n",
       "      <td>7.01</td>\n",
       "      <td>4.96</td>\n",
       "      <td>...</td>\n",
       "      <td>49.11</td>\n",
       "      <td>23.74</td>\n",
       "      <td>23.74</td>\n",
       "      <td>16.93</td>\n",
       "      <td>31.32</td>\n",
       "      <td>30.11</td>\n",
       "      <td>30.11</td>\n",
       "      <td>0.82</td>\n",
       "      <td>0.82</td>\n",
       "      <td>1.17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DNMT3A_P904Q_ch</td>\n",
       "      <td>12.28</td>\n",
       "      <td>13.90</td>\n",
       "      <td>22.69</td>\n",
       "      <td>11.32</td>\n",
       "      <td>31.15</td>\n",
       "      <td>11.51</td>\n",
       "      <td>3.75</td>\n",
       "      <td>9.87</td>\n",
       "      <td>5.75</td>\n",
       "      <td>...</td>\n",
       "      <td>52.89</td>\n",
       "      <td>41.59</td>\n",
       "      <td>41.59</td>\n",
       "      <td>33.98</td>\n",
       "      <td>46.25</td>\n",
       "      <td>37.95</td>\n",
       "      <td>37.95</td>\n",
       "      <td>0.81</td>\n",
       "      <td>0.89</td>\n",
       "      <td>1.19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DNMT3A_P904Q_refseq</td>\n",
       "      <td>8.04</td>\n",
       "      <td>7.35</td>\n",
       "      <td>0.99</td>\n",
       "      <td>0.67</td>\n",
       "      <td>25.15</td>\n",
       "      <td>9.09</td>\n",
       "      <td>3.76</td>\n",
       "      <td>7.01</td>\n",
       "      <td>4.96</td>\n",
       "      <td>...</td>\n",
       "      <td>49.11</td>\n",
       "      <td>23.74</td>\n",
       "      <td>23.74</td>\n",
       "      <td>16.93</td>\n",
       "      <td>31.32</td>\n",
       "      <td>30.11</td>\n",
       "      <td>30.11</td>\n",
       "      <td>0.82</td>\n",
       "      <td>0.82</td>\n",
       "      <td>1.17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DNMT3A_P904R_ch</td>\n",
       "      <td>11.95</td>\n",
       "      <td>13.49</td>\n",
       "      <td>26.51</td>\n",
       "      <td>14.69</td>\n",
       "      <td>30.02</td>\n",
       "      <td>10.98</td>\n",
       "      <td>3.43</td>\n",
       "      <td>10.44</td>\n",
       "      <td>6.18</td>\n",
       "      <td>...</td>\n",
       "      <td>51.32</td>\n",
       "      <td>53.38</td>\n",
       "      <td>53.38</td>\n",
       "      <td>42.37</td>\n",
       "      <td>42.22</td>\n",
       "      <td>49.26</td>\n",
       "      <td>49.26</td>\n",
       "      <td>1.21</td>\n",
       "      <td>1.28</td>\n",
       "      <td>2.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103</th>\n",
       "      <td>SRSF2_P95R_refseq</td>\n",
       "      <td>11.10</td>\n",
       "      <td>36.72</td>\n",
       "      <td>19.38</td>\n",
       "      <td>3.89</td>\n",
       "      <td>33.17</td>\n",
       "      <td>20.87</td>\n",
       "      <td>12.42</td>\n",
       "      <td>6.89</td>\n",
       "      <td>16.39</td>\n",
       "      <td>...</td>\n",
       "      <td>42.34</td>\n",
       "      <td>5.37</td>\n",
       "      <td>5.37</td>\n",
       "      <td>2.43</td>\n",
       "      <td>12.47</td>\n",
       "      <td>0.84</td>\n",
       "      <td>0.84</td>\n",
       "      <td>2.82</td>\n",
       "      <td>1.04</td>\n",
       "      <td>2.18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104</th>\n",
       "      <td>TP53_R175H_ch</td>\n",
       "      <td>13.89</td>\n",
       "      <td>19.00</td>\n",
       "      <td>30.48</td>\n",
       "      <td>29.72</td>\n",
       "      <td>24.22</td>\n",
       "      <td>53.28</td>\n",
       "      <td>10.49</td>\n",
       "      <td>11.97</td>\n",
       "      <td>39.93</td>\n",
       "      <td>...</td>\n",
       "      <td>50.00</td>\n",
       "      <td>36.00</td>\n",
       "      <td>36.00</td>\n",
       "      <td>34.15</td>\n",
       "      <td>58.28</td>\n",
       "      <td>49.46</td>\n",
       "      <td>49.46</td>\n",
       "      <td>13.89</td>\n",
       "      <td>22.95</td>\n",
       "      <td>10.08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105</th>\n",
       "      <td>TP53_R175H_refseq</td>\n",
       "      <td>14.10</td>\n",
       "      <td>17.70</td>\n",
       "      <td>54.38</td>\n",
       "      <td>27.30</td>\n",
       "      <td>22.04</td>\n",
       "      <td>53.69</td>\n",
       "      <td>9.09</td>\n",
       "      <td>11.85</td>\n",
       "      <td>53.04</td>\n",
       "      <td>...</td>\n",
       "      <td>49.16</td>\n",
       "      <td>35.86</td>\n",
       "      <td>35.86</td>\n",
       "      <td>22.26</td>\n",
       "      <td>44.61</td>\n",
       "      <td>39.17</td>\n",
       "      <td>39.17</td>\n",
       "      <td>16.28</td>\n",
       "      <td>24.90</td>\n",
       "      <td>13.51</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106</th>\n",
       "      <td>TP53_R273H_ch</td>\n",
       "      <td>3.52</td>\n",
       "      <td>54.74</td>\n",
       "      <td>12.27</td>\n",
       "      <td>16.16</td>\n",
       "      <td>28.71</td>\n",
       "      <td>36.98</td>\n",
       "      <td>5.89</td>\n",
       "      <td>2.62</td>\n",
       "      <td>21.58</td>\n",
       "      <td>...</td>\n",
       "      <td>49.16</td>\n",
       "      <td>17.57</td>\n",
       "      <td>17.57</td>\n",
       "      <td>8.10</td>\n",
       "      <td>24.17</td>\n",
       "      <td>43.39</td>\n",
       "      <td>43.39</td>\n",
       "      <td>12.18</td>\n",
       "      <td>24.21</td>\n",
       "      <td>15.39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>TP53_R273H_refseq</td>\n",
       "      <td>7.87</td>\n",
       "      <td>8.38</td>\n",
       "      <td>1.36</td>\n",
       "      <td>1.92</td>\n",
       "      <td>10.27</td>\n",
       "      <td>48.61</td>\n",
       "      <td>9.22</td>\n",
       "      <td>5.56</td>\n",
       "      <td>41.64</td>\n",
       "      <td>...</td>\n",
       "      <td>51.84</td>\n",
       "      <td>12.50</td>\n",
       "      <td>12.50</td>\n",
       "      <td>7.41</td>\n",
       "      <td>22.63</td>\n",
       "      <td>47.00</td>\n",
       "      <td>47.00</td>\n",
       "      <td>9.43</td>\n",
       "      <td>20.39</td>\n",
       "      <td>12.41</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>108 rows × 594 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "HLA_formatted          gene_var_gt  DPA1_103-DPB1_1001  DPA1_103-DPB1_101  \\\n",
       "0                  DNMT3A_P904L_ch               20.84               7.30   \n",
       "1              DNMT3A_P904L_refseq                8.04               7.35   \n",
       "2                  DNMT3A_P904Q_ch               12.28              13.90   \n",
       "3              DNMT3A_P904Q_refseq                8.04               7.35   \n",
       "4                  DNMT3A_P904R_ch               11.95              13.49   \n",
       "..                             ...                 ...                ...   \n",
       "103              SRSF2_P95R_refseq               11.10              36.72   \n",
       "104                  TP53_R175H_ch               13.89              19.00   \n",
       "105              TP53_R175H_refseq               14.10              17.70   \n",
       "106                  TP53_R273H_ch                3.52              54.74   \n",
       "107              TP53_R273H_refseq                7.87               8.38   \n",
       "\n",
       "HLA_formatted  DPA1_103-DPB1_1101  DPA1_103-DPB1_1301  DPA1_103-DPB1_1401  \\\n",
       "0                            6.75                5.05               36.28   \n",
       "1                            0.99                0.67               25.15   \n",
       "2                           22.69               11.32               31.15   \n",
       "3                            0.99                0.67               25.15   \n",
       "4                           26.51               14.69               30.02   \n",
       "..                            ...                 ...                 ...   \n",
       "103                         19.38                3.89               33.17   \n",
       "104                         30.48               29.72               24.22   \n",
       "105                         54.38               27.30               22.04   \n",
       "106                         12.27               16.16               28.71   \n",
       "107                          1.36                1.92               10.27   \n",
       "\n",
       "HLA_formatted  DPA1_103-DPB1_1501  DPA1_103-DPB1_1601  DPA1_103-DPB1_1701  \\\n",
       "0                           21.29                8.93               17.73   \n",
       "1                            9.09                3.76                7.01   \n",
       "2                           11.51                3.75                9.87   \n",
       "3                            9.09                3.76                7.01   \n",
       "4                           10.98                3.43               10.44   \n",
       "..                            ...                 ...                 ...   \n",
       "103                         20.87               12.42                6.89   \n",
       "104                         53.28               10.49               11.97   \n",
       "105                         53.69                9.09               11.85   \n",
       "106                         36.98                5.89                2.62   \n",
       "107                         48.61                9.22                5.56   \n",
       "\n",
       "HLA_formatted  DPA1_103-DPB1_1801  ...  DRB3_101  DRB3_202  DRB3_210  \\\n",
       "0                           17.99  ...     64.62     54.75     54.75   \n",
       "1                            4.96  ...     49.11     23.74     23.74   \n",
       "2                            5.75  ...     52.89     41.59     41.59   \n",
       "3                            4.96  ...     49.11     23.74     23.74   \n",
       "4                            6.18  ...     51.32     53.38     53.38   \n",
       "..                            ...  ...       ...       ...       ...   \n",
       "103                         16.39  ...     42.34      5.37      5.37   \n",
       "104                         39.93  ...     50.00     36.00     36.00   \n",
       "105                         53.04  ...     49.16     35.86     35.86   \n",
       "106                         21.58  ...     49.16     17.57     17.57   \n",
       "107                         41.64  ...     51.84     12.50     12.50   \n",
       "\n",
       "HLA_formatted  DRB3_224  DRB3_301  DRB4_101  DRB4_103  DRB5_101  DRB5_102  \\\n",
       "0                 47.57     60.38     57.86     57.86      1.54      1.54   \n",
       "1                 16.93     31.32     30.11     30.11      0.82      0.82   \n",
       "2                 33.98     46.25     37.95     37.95      0.81      0.89   \n",
       "3                 16.93     31.32     30.11     30.11      0.82      0.82   \n",
       "4                 42.37     42.22     49.26     49.26      1.21      1.28   \n",
       "..                  ...       ...       ...       ...       ...       ...   \n",
       "103                2.43     12.47      0.84      0.84      2.82      1.04   \n",
       "104               34.15     58.28     49.46     49.46     13.89     22.95   \n",
       "105               22.26     44.61     39.17     39.17     16.28     24.90   \n",
       "106                8.10     24.17     43.39     43.39     12.18     24.21   \n",
       "107                7.41     22.63     47.00     47.00      9.43     20.39   \n",
       "\n",
       "HLA_formatted  DRB5_202  \n",
       "0                  2.05  \n",
       "1                  1.17  \n",
       "2                  1.19  \n",
       "3                  1.17  \n",
       "4                  2.00  \n",
       "..                  ...  \n",
       "103                2.18  \n",
       "104               10.08  \n",
       "105               13.51  \n",
       "106               15.39  \n",
       "107               12.41  \n",
       "\n",
       "[108 rows x 594 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# create a suitably formatted NetMHC df \n",
    "param = '%Rank_EL'\n",
    "pred_df_sub = pred_df_all[['HLA_formatted', 'gene_var_gt', param]]\n",
    "pred_df_sub_wide = pd.pivot(pred_df_sub, index='gene_var_gt', columns='HLA_formatted', values=param)\n",
    "pred_df_sub_wide = pred_df_sub_wide.reset_index() # this is to make sure that you have the gene_var column in there too \n",
    "pred_df_sub_wide"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['DRB5_101', 'DRB5_102', 'DRB5_202', 'DRB5_9901', 'DRB4_101', 'DRB4_103',\n",
       "       'DRB4_9901', 'DRB3_101', 'DRB3_202', 'DRB3_210',\n",
       "       ...\n",
       "       'DQA1_102-DQB1_609', 'DQA1_103-DQB1_609', 'DQA1_201-DQB1_609',\n",
       "       'DQA1_301-DQB1_609', 'DQA1_302-DQB1_609', 'DQA1_303-DQB1_609',\n",
       "       'DQA1_401-DQB1_609', 'DQA1_501-DQB1_609', 'DQA1_505-DQB1_609',\n",
       "       'DQA1_601-DQB1_609'],\n",
       "      dtype='object', length=446)"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hla_ukbb_all = df_hla2_drpq.columns[22:] # select columns with alleles \n",
    "hla_ukbb_all"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create a list of HLA alleles genotyped in the UKBB for which predictions are available\n",
    "hla_intersect = pred_df_sub_wide.columns[pred_df_sub_wide.columns.isin(hla_ukbb_all)].tolist() # DRB alleles \n",
    "hla_intersect = list(set(hla_intersect) & set(hla_ukbb_all))\n",
    "\n",
    "# # create the prediction dataset \n",
    "pred_df_sub = pred_df_sub_wide[hla_intersect + pred_df_sub_wide.columns[pred_df_sub_wide.columns.str.contains('gene_var')].tolist()] # subset netmhc so you only have alleles which are in the UKBB, also 194\n",
    "pred_df_sub = pred_df_sub[pred_df_sub['gene_var_gt'].str.contains('_ch', regex=True)] # retain CH scores only \n",
    "pred_df_sub['gene_var'] = pred_df_sub['gene_var_gt'].str.replace('_ch', '') # remove the ch / refseq annotation\n",
    "pred_df_sub['gene_var'] = pred_df_sub['gene_var'].str.replace('_refseq', '') # remove refseq if present \n",
    "\n",
    "# apply the function to find, for each participant, the best score for each examined CH variant \n",
    "# you take your current df and apply, row by row, the function to get scores for each variant \n",
    "df_hla2_hlas_drpq = pd.concat([df_hla2_drpq[['Person_ID', 'gene_var', 'VAF', 'var_depth', 'depth', 'ch_status', 'age']], df_hla2_drpq[hla_intersect]], axis = 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# apply the function to find scores for each variant \n",
    "df_hla2_scores = pd.concat([df_hla2_hlas_drpq, df_hla2_hlas_drpq.apply(find_best_score_for_all_variants, df=pred_df_sub, param=param, axis=1)], axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Person_ID</th>\n",
       "      <th>gene_var</th>\n",
       "      <th>VAF</th>\n",
       "      <th>var_depth</th>\n",
       "      <th>depth</th>\n",
       "      <th>ch_status</th>\n",
       "      <th>age</th>\n",
       "      <th>DRB1_408</th>\n",
       "      <th>DQA1_301-DQB1_302</th>\n",
       "      <th>DQA1_101-DQB1_605</th>\n",
       "      <th>...</th>\n",
       "      <th>score_TP53_R175H</th>\n",
       "      <th>score_TP53_R273H</th>\n",
       "      <th>het_allele_II_DRB</th>\n",
       "      <th>het_allele_II_DPA</th>\n",
       "      <th>het_allele_II_DPB</th>\n",
       "      <th>het_allele_II_DQA</th>\n",
       "      <th>het_allele_II_DQB</th>\n",
       "      <th>count_class_II</th>\n",
       "      <th>sum_class_II</th>\n",
       "      <th>het_all_class_II</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>4860169</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>46</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>7.91</td>\n",
       "      <td>3.15</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>14</td>\n",
       "      <td>16.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3381323</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>52</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>20.19</td>\n",
       "      <td>21.06</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>13</td>\n",
       "      <td>16.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2805252</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>65</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2.89</td>\n",
       "      <td>3.15</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>15</td>\n",
       "      <td>16.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3318036</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>50</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>18.58</td>\n",
       "      <td>23.87</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>8</td>\n",
       "      <td>16.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4120291</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>41</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>4.19</td>\n",
       "      <td>4.71</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>13</td>\n",
       "      <td>16.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>328236</th>\n",
       "      <td>4478244</td>\n",
       "      <td>DNMT3A_P904Q</td>\n",
       "      <td>0.018692</td>\n",
       "      <td>2.0</td>\n",
       "      <td>107.0</td>\n",
       "      <td>1</td>\n",
       "      <td>55</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>6.47</td>\n",
       "      <td>3.15</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>14</td>\n",
       "      <td>16.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>328237</th>\n",
       "      <td>4478244</td>\n",
       "      <td>DNMT3A_R882H</td>\n",
       "      <td>0.022727</td>\n",
       "      <td>2.0</td>\n",
       "      <td>88.0</td>\n",
       "      <td>1</td>\n",
       "      <td>55</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>6.47</td>\n",
       "      <td>3.15</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>14</td>\n",
       "      <td>16.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>328238</th>\n",
       "      <td>3573995</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>62</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>11.96</td>\n",
       "      <td>17.57</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>9</td>\n",
       "      <td>16.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>328239</th>\n",
       "      <td>3025735</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>51</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>14.15</td>\n",
       "      <td>8.58</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>14</td>\n",
       "      <td>16.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>328240</th>\n",
       "      <td>1381182</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>67</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>7.98</td>\n",
       "      <td>3.15</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>15</td>\n",
       "      <td>16.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>328241 rows × 508 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Person_ID      gene_var       VAF  var_depth  depth  ch_status  age  \\\n",
       "0         4860169           NaN       NaN        NaN    NaN          0   46   \n",
       "1         3381323           NaN       NaN        NaN    NaN          0   52   \n",
       "2         2805252           NaN       NaN        NaN    NaN          0   65   \n",
       "3         3318036           NaN       NaN        NaN    NaN          0   50   \n",
       "4         4120291           NaN       NaN        NaN    NaN          0   41   \n",
       "...           ...           ...       ...        ...    ...        ...  ...   \n",
       "328236    4478244  DNMT3A_P904Q  0.018692        2.0  107.0          1   55   \n",
       "328237    4478244  DNMT3A_R882H  0.022727        2.0   88.0          1   55   \n",
       "328238    3573995           NaN       NaN        NaN    NaN          0   62   \n",
       "328239    3025735           NaN       NaN        NaN    NaN          0   51   \n",
       "328240    1381182           NaN       NaN        NaN    NaN          0   67   \n",
       "\n",
       "        DRB1_408  DQA1_301-DQB1_302  DQA1_101-DQB1_605  ...  score_TP53_R175H  \\\n",
       "0            0.0                0.0                0.0  ...              7.91   \n",
       "1            0.0                1.0                0.0  ...             20.19   \n",
       "2            0.0                0.0                0.0  ...              2.89   \n",
       "3            0.0                0.0                0.0  ...             18.58   \n",
       "4            0.0                0.0                0.0  ...              4.19   \n",
       "...          ...                ...                ...  ...               ...   \n",
       "328236       0.0                0.0                0.0  ...              6.47   \n",
       "328237       0.0                0.0                0.0  ...              6.47   \n",
       "328238       0.0                0.0                0.0  ...             11.96   \n",
       "328239       0.0                0.0                0.0  ...             14.15   \n",
       "328240       0.0                0.0                0.0  ...              7.98   \n",
       "\n",
       "        score_TP53_R273H  het_allele_II_DRB  het_allele_II_DPA  \\\n",
       "0                   3.15               True              False   \n",
       "1                  21.06               True              False   \n",
       "2                   3.15               True               True   \n",
       "3                  23.87              False              False   \n",
       "4                   4.71               True              False   \n",
       "...                  ...                ...                ...   \n",
       "328236              3.15               True              False   \n",
       "328237              3.15               True              False   \n",
       "328238             17.57              False              False   \n",
       "328239              8.58               True              False   \n",
       "328240              3.15               True               True   \n",
       "\n",
       "        het_allele_II_DPB  het_allele_II_DQA  het_allele_II_DQB  \\\n",
       "0                    True               True               True   \n",
       "1                    True               True               True   \n",
       "2                    True               True               True   \n",
       "3                   False              False              False   \n",
       "4                    True               True               True   \n",
       "...                   ...                ...                ...   \n",
       "328236               True               True               True   \n",
       "328237               True               True               True   \n",
       "328238              False              False              False   \n",
       "328239               True               True               True   \n",
       "328240               True               True               True   \n",
       "\n",
       "        count_class_II  sum_class_II  het_all_class_II  \n",
       "0                   14          16.0             False  \n",
       "1                   13          16.0             False  \n",
       "2                   15          16.0             False  \n",
       "3                    8          16.0             False  \n",
       "4                   13          16.0             False  \n",
       "...                ...           ...               ...  \n",
       "328236              14          16.0             False  \n",
       "328237              14          16.0             False  \n",
       "328238               9          16.0             False  \n",
       "328239              14          16.0             False  \n",
       "328240              15          16.0             False  \n",
       "\n",
       "[328241 rows x 508 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# add extra columns (MHC data)\n",
    "df_hla2_scores_added = pd.concat([df_hla2_scores, df_hla2[['het_allele_II_DRB', 'het_allele_II_DPA', 'het_allele_II_DPB', 'het_allele_II_DQA', 'het_allele_II_DQB', 'count_class_II', 'sum_class_II', 'het_all_class_II']]], axis = 1)\n",
    "df_hla2_scores_added"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "# save to a file (this is already usable for further analysis)\n",
    "df_hla2_scores_added.to_csv('/Users/barbarawalkowiak/Desktop/msc_thesis/results/dataframes/20240907_netmhc2_scores_for_all_var.csv')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# read in the dataframe \n",
    "df_hla2_scores_added = pd.read_csv('/Users/barbarawalkowiak/Desktop/msc_thesis/results/dataframes/20240907_netmhc2_scores_for_all_var.csv')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Add label (top / bottom half of binding scores)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "# subset the dataframe to only include gene_var, Person ID and scores \n",
    "scores_col = [col for col in df_hla2_scores_added.columns if col.startswith('score_')]\n",
    "df_hla2_scores_sub = pd.concat([df_hla2_scores_added[['Person_ID', 'gene_var', 'var_depth', 'depth', 'VAF']], df_hla2_scores_added[scores_col]], axis = 1)\n",
    "\n",
    "# melt the dataframe \n",
    "df_hla2_scores_sub_melted = pd.melt(df_hla2_scores_sub, id_vars = ['gene_var', 'Person_ID', 'var_depth', 'depth', 'VAF'])\n",
    "\n",
    "# add a column which has the name of the variant (nicely formatted)\n",
    "df_hla2_scores_sub_melted['CH_variant'] = df_hla2_scores_sub_melted['variable'].str[6:]\n",
    "\n",
    "# add a column to indicate CH status (either carrier of the variant the score is for, or non-carrier, even if have CH driven by sth else)\n",
    "df_hla2_scores_sub_melted['CH_status'] = np.where(df_hla2_scores_sub_melted['gene_var'] == df_hla2_scores_sub_melted['CH_variant'].replace(), 1, 0)\n",
    "\n",
    "# you want the %Rank_EL to be in the format of -1*log10(%Rank_EL)\n",
    "df_hla2_scores_sub_melted['log_score'] = -1 * np.log10(df_hla2_scores_sub_melted['value'])\n",
    "\n",
    "# add median score for each variant and then order them by median \n",
    "df_hla2_scores_sub_melted['median_score'] = df_hla2_scores_sub_melted.groupby('CH_variant')['log_score'].transform('median')\n",
    "\n",
    "# convert to categories \n",
    "df_hla2_scores_sub_melted['gene_var'] = df_hla2_scores_sub_melted['gene_var'].astype('category') \n",
    "df_hla2_scores_sub_melted['CH_variant'] = df_hla2_scores_sub_melted['CH_variant'].astype('category') # make sure you convert this to category first of all"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add group (split participants with values equal to the median)\n",
    "\n",
    "variants = df_hla2_scores_sub_melted.CH_variant.unique().tolist()\n",
    "\n",
    "for var in variants:\n",
    "\n",
    "    # select df with variant \n",
    "    df_variant = df_hla2_scores_sub_melted[df_hla2_scores_sub_melted['CH_variant'] == var]\n",
    "    \n",
    "    # get the median \n",
    "    median_score = df_variant['log_score'].median()\n",
    "    \n",
    "    # find median values \n",
    "    median_values = df_variant[df_variant['log_score'] == df_variant['median_score']]\n",
    "    \n",
    "    # assign top / bottom if above or below median \n",
    "    below_median = df_variant[df_variant['log_score'] < median_score]\n",
    "    above_median = df_variant[df_variant['log_score'] > median_score]\n",
    "    \n",
    "    # figure out how many observations you need with median values to top / bottom half\n",
    "    half_length = len(df_variant) // 2\n",
    "    num_bottom_needed = half_length - len(below_median)\n",
    "    num_top_needed = len(median_values) - num_bottom_needed\n",
    "    \n",
    "    # assign values equal to median to top or bottom\n",
    "    shuffled_median_values = median_values.sample(frac=1)\n",
    "    bottom_half_median = shuffled_median_values.iloc[:num_bottom_needed]\n",
    "    top_half_median = shuffled_median_values.iloc[num_bottom_needed:]\n",
    "    \n",
    "    # assign groups\n",
    "    df_hla2_scores_sub_melted.loc[below_median.index, 'group'] = 'bottom half'\n",
    "    df_hla2_scores_sub_melted.loc[above_median.index, 'group'] = 'top half'\n",
    "    df_hla2_scores_sub_melted.loc[bottom_half_median.index, 'group'] = 'bottom half'\n",
    "    df_hla2_scores_sub_melted.loc[top_half_median.index, 'group'] = 'top half'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "ename": "",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31mThe Kernel crashed while executing code in the current cell or a previous cell. \n",
      "\u001b[1;31mPlease review the code in the cell(s) to identify a possible cause of the failure. \n",
      "\u001b[1;31mClick <a href='https://aka.ms/vscodeJupyterKernelCrash'>here</a> for more info. \n",
      "\u001b[1;31mView Jupyter <a href='command:jupyter.viewOutput'>log</a> for further details."
     ]
    }
   ],
   "source": [
    "# modified saving labels\n",
    "\n",
    "# save to a file (this is already usable for further analysis)\n",
    "df_hla2_scores_sub_melted.to_csv('/Users/barbarawalkowiak/Desktop/msc_thesis/results/dataframes/20240907_netmhc2_scores_for_all_var_with_labels.csv')\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
