# MHC_ClonalHaematopoiesis

PACKAGE VERSIONS / SOFTWARE

- NetMHCpan-4.1
- NEtMHCIIpan-4.3
- PRIME2.0

- Jupyter Notebook 6.5.4
- Python 3.11.5
- Pandas 2.0.3
- Numpy 1.24.3
- Seaborn 0.13.2
- Matplotlib 3.7.2

SCRIPTS 
1. Generate kmers of proteins with CH variants of appropriate length.
2. Obtain predictions for kmer-MHC allele binding
3. Obtain datasframes with prediction scores
4. Obtain summary dataframes (choose best variant-MHC combination)
5. Obtain dataframe with MHC genotype and CH status for UKB participants
6. Add MHC binding scores to each participant examined
7. (A) Generate main figures (1-5), supplementary figures (1, 2, 4, 5, 7, 9, 10-13 (run with PRIME predictions), 18, 19), (B) Generate supplementary figures 14-17
8. Generate supplementary figure 6
9. Run and analyse power simulations (supplementary figures 3 and 8) 

